{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- cluster sequences based on Ag1000g and sequencing data thresholds\n",
    "- compare clustering performance (Ag1000g thresholds have higher resolution)\n",
    "- filter outlier samples within species (single genotype in a separate cluster)\n",
    "- filter alleles split between clusters\n",
    "\n",
    "Manual steps: none"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run common.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bad_samples = pd.read_csv(BAD_SAMPLES).iloc[:, 0].to_list()\n",
    "len(bad_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_Sample</th>\n",
       "      <th>target</th>\n",
       "      <th>consensus</th>\n",
       "      <th>reads</th>\n",
       "      <th>species</th>\n",
       "      <th>combUID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abro-21</td>\n",
       "      <td>13</td>\n",
       "      <td>TCTCCATTTCTATCTGATGAAACTTGTGCCCGTGGTTTAATTAATG...</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Anopheles_brohieri</td>\n",
       "      <td>13-0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  s_Sample  target                                          consensus  reads  \\\n",
       "0  Abro-21      13  TCTCCATTTCTATCTGATGAAACTTGTGCCCGTGGTTTAATTAATG...   23.0   \n",
       "\n",
       "              species combUID  \n",
       "0  Anopheles_brohieri    13-0  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_data = pd.read_csv(COMB_SEQ)\n",
    "# remove species ID for compromised samples\n",
    "seq_data.loc[seq_data.s_Sample.isin(bad_samples), 'species'] = 'unknown'\n",
    "seq_data.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>start_insert</th>\n",
       "      <th>end_insert</th>\n",
       "      <th>len_insert</th>\n",
       "      <th>var_ag1k</th>\n",
       "      <th>dist_ag1k</th>\n",
       "      <th>aln_len</th>\n",
       "      <th>ngaps_mean</th>\n",
       "      <th>var_multisp</th>\n",
       "      <th>delta_t</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>10356439</td>\n",
       "      <td>10356587</td>\n",
       "      <td>149</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.033557</td>\n",
       "      <td>233</td>\n",
       "      <td>95.275362</td>\n",
       "      <td>12.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   target  start_insert  end_insert  len_insert  var_ag1k  dist_ag1k  aln_len  \\\n",
       "0       0      10356439    10356587         149       5.0   0.033557      233   \n",
       "\n",
       "   ngaps_mean  var_multisp  delta_t  \n",
       "0   95.275362         12.0      7.0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t_data = pd.read_csv(COMB_THRESHOLDS)\n",
    "t_data.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cd-hit-est clustering with two sets of identity thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9914529914529915"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def clust(ampl, max_var=1, word_len=10):\n",
    "    \n",
    "    # subset amplicon data\n",
    "    ampl_data = seq_data[seq_data.target == ampl]\n",
    "    # get only unique sequences\n",
    "    ampl_data = ampl_data[~ampl_data.combUID.duplicated()]\n",
    "    # write sequences under combUID name\n",
    "    with open('temp.fa', 'w') as o:\n",
    "        for (i, row) in ampl_data.iterrows():\n",
    "            o.write('>{}\\n{}\\n'.format(row.combUID,\n",
    "                                       row.consensus))\n",
    "    \n",
    "    # cd-hit-est identity definition:\n",
    "    # number of identical bases in alignment\n",
    "    # divided by the full length of the shorter sequence\n",
    "    min_len = ampl_data.consensus.str.len().min()\n",
    "    identity = 1 - max_var / float(min_len)\n",
    "    # cd-hit does not work for <0.8 identity\n",
    "    identity = max(identity, 0.8)\n",
    "    \n",
    "    # run clustering\n",
    "    ! cd-hit-est -i temp.fa -o cl -c {identity} -n {word_len} > cl.log\n",
    "    # parse clusters into dict\n",
    "    cl_data = dict()\n",
    "    with open('cl.clstr') as f:\n",
    "        for l in f:\n",
    "            l = l.strip('\\n')\n",
    "            # cluster name\n",
    "            if l.startswith('>'):\n",
    "                cluster = 'cl-{}-{}'.format(ampl,\n",
    "                                            l.split(' ')[1])\n",
    "            # cluster members\n",
    "            else:\n",
    "                seq = l.split('>')[1].split('...')[0]\n",
    "                cl_data[seq] = cluster\n",
    "    # clean up\n",
    "    ! rm cl cl.clstr cl.log temp.fa\n",
    "    # all input sequences clustered?\n",
    "    assert len(cl_data) == ampl_data.combUID.nunique()\n",
    "    \n",
    "    return cl_data, identity\n",
    "clust(0)[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61\n",
      "Done!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "5606"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "{0: 0.9572649572649573,\n",
       " 1: 0.9869281045751634,\n",
       " 2: 0.9482758620689655,\n",
       " 3: 0.9230769230769231,\n",
       " 4: 0.9789473684210527,\n",
       " 5: 0.9465648854961832,\n",
       " 6: 0.9615384615384616,\n",
       " 7: 0.9507042253521126,\n",
       " 8: 0.9541984732824428,\n",
       " 9: 0.9735099337748344,\n",
       " 10: 0.9618320610687023,\n",
       " 11: 0.9416666666666667,\n",
       " 12: 0.9181818181818182,\n",
       " 13: 0.9669421487603306,\n",
       " 14: 0.9748427672955975,\n",
       " 15: 0.9748427672955975,\n",
       " 16: 0.9761904761904762,\n",
       " 17: 0.9612903225806452,\n",
       " 18: 0.9624060150375939,\n",
       " 19: 0.9722222222222222,\n",
       " 20: 0.9792746113989638,\n",
       " 21: 0.9728260869565217,\n",
       " 22: 0.9577464788732395,\n",
       " 23: 0.9278350515463918,\n",
       " 24: 0.9808917197452229,\n",
       " 25: 0.9212598425196851,\n",
       " 26: 0.8876404494382022,\n",
       " 27: 0.9820359281437125,\n",
       " 28: 0.9797297297297297,\n",
       " 29: 0.9861111111111112,\n",
       " 30: 0.9849624060150376,\n",
       " 31: 0.9666666666666667,\n",
       " 32: 0.875,\n",
       " 33: 0.8901098901098901,\n",
       " 34: 0.9743589743589743,\n",
       " 35: 0.9473684210526316,\n",
       " 36: 0.974025974025974,\n",
       " 37: 0.952755905511811,\n",
       " 38: 0.9456521739130435,\n",
       " 39: 0.9814814814814815,\n",
       " 40: 0.9411764705882353,\n",
       " 41: 0.9594594594594594,\n",
       " 42: 0.941747572815534,\n",
       " 43: 0.9558823529411765,\n",
       " 44: 0.9333333333333333,\n",
       " 45: 0.9595959595959596,\n",
       " 46: 0.9186046511627907,\n",
       " 47: 0.9615384615384616,\n",
       " 48: 0.9310344827586207,\n",
       " 49: 0.945054945054945,\n",
       " 50: 0.9603960396039604,\n",
       " 51: 0.9541984732824428,\n",
       " 52: 0.9389312977099237,\n",
       " 53: 0.9836065573770492,\n",
       " 54: 0.9545454545454546,\n",
       " 55: 0.9615384615384616,\n",
       " 56: 0.9793103448275862,\n",
       " 57: 0.9803921568627451,\n",
       " 58: 0.9876543209876543,\n",
       " 59: 0.9230769230769231,\n",
       " 60: 0.9523809523809523,\n",
       " 61: 0.9873417721518988}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ag1k thresholds\n",
    "gc_clustering = dict()\n",
    "gc_identities = dict()\n",
    "for ampl in AMPLS:\n",
    "    sys.stdout.write('\\r'+ampl)\n",
    "    ampl = int(ampl)\n",
    "    # variants\n",
    "    v = t_data.loc[t_data.target == ampl, 'var_ag1k'].iloc[0]\n",
    "    # clustering\n",
    "    (c, i) = clust(ampl, v)\n",
    "    # record\n",
    "    gc_clustering.update(c)\n",
    "    gc_identities[ampl] = i\n",
    "print('\\nDone!')\n",
    "display(len(gc_clustering))\n",
    "gc_identities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "61\n",
      "Done!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "5606"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "{0: 0.8974358974358975,\n",
       " 1: 0.9803921568627451,\n",
       " 2: 0.9051724137931034,\n",
       " 3: 0.9038461538461539,\n",
       " 4: 0.9894736842105263,\n",
       " 5: 0.9312977099236641,\n",
       " 6: 0.9,\n",
       " 7: 0.8802816901408451,\n",
       " 8: 0.8931297709923665,\n",
       " 9: 0.9801324503311258,\n",
       " 10: 0.8244274809160306,\n",
       " 11: 0.8833333333333333,\n",
       " 12: 0.9454545454545454,\n",
       " 13: 0.9173553719008265,\n",
       " 14: 0.9685534591194969,\n",
       " 15: 0.9685534591194969,\n",
       " 16: 0.8988095238095238,\n",
       " 17: 0.832258064516129,\n",
       " 18: 0.9097744360902256,\n",
       " 19: 0.9305555555555556,\n",
       " 20: 0.9792746113989638,\n",
       " 21: 0.9836956521739131,\n",
       " 22: 0.9577464788732395,\n",
       " 23: 0.8,\n",
       " 24: 0.9808917197452229,\n",
       " 25: 0.889763779527559,\n",
       " 26: 0.8089887640449438,\n",
       " 27: 0.9640718562874252,\n",
       " 28: 0.9527027027027027,\n",
       " 29: 0.9861111111111112,\n",
       " 30: 0.8,\n",
       " 31: 0.9809523809523809,\n",
       " 32: 0.8,\n",
       " 33: 0.8571428571428572,\n",
       " 34: 0.9807692307692307,\n",
       " 35: 0.9532163742690059,\n",
       " 36: 0.974025974025974,\n",
       " 37: 0.8582677165354331,\n",
       " 38: 0.9456521739130435,\n",
       " 39: 0.9814814814814815,\n",
       " 40: 0.8,\n",
       " 41: 0.9054054054054054,\n",
       " 42: 0.8640776699029127,\n",
       " 43: 0.8602941176470589,\n",
       " 44: 0.8,\n",
       " 45: 0.8282828282828283,\n",
       " 46: 0.8604651162790697,\n",
       " 47: 0.8653846153846154,\n",
       " 48: 0.8,\n",
       " 49: 0.8,\n",
       " 50: 0.8,\n",
       " 51: 0.8549618320610687,\n",
       " 52: 0.916030534351145,\n",
       " 53: 0.9890710382513661,\n",
       " 54: 0.8863636363636364,\n",
       " 55: 0.8,\n",
       " 56: 0.9862068965517241,\n",
       " 57: 0.9803921568627451,\n",
       " 58: 0.9814814814814815,\n",
       " 59: 0.8,\n",
       " 60: 0.8968253968253969,\n",
       " 61: 0.9810126582278481}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# multispecies thresholds\n",
    "mu_clustering = dict()\n",
    "mu_identities = dict()\n",
    "for ampl in AMPLS:\n",
    "    sys.stdout.write('\\r'+ampl)\n",
    "    ampl = int(ampl)\n",
    "    # variants\n",
    "    v = t_data.loc[t_data.target == ampl, 'var_multisp'].iloc[0]\n",
    "    # clustering\n",
    "    (c, i) = clust(ampl, v)\n",
    "    # record\n",
    "    mu_clustering.update(c)\n",
    "    mu_identities[ampl] = i\n",
    "print('\\nDone!')\n",
    "display(len(mu_clustering))\n",
    "mu_identities"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In case of multispecies-based clustering, minimum 80% similarity threshold was hit 7 times."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare clustering performance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine clustering results with sequencing data\n",
    "seq_data['ag1k_cluster'] = seq_data.combUID.replace(gc_clustering)\n",
    "seq_data['multisp_cluster'] = seq_data.combUID.replace(mu_clustering)\n",
    "clustering_types = ('ag1k_cluster', 'multisp_cluster')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10057, 8)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(9589, 8)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# limit stats calculation to high quality species\n",
    "hq_seq_data = seq_data[seq_data.species != 'unknown']\n",
    "display(seq_data.shape, hq_seq_data.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ag1k_cluster</th>\n",
       "      <th>multisp_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1951.000000</td>\n",
       "      <td>1353.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1.667350</td>\n",
       "      <td>2.254989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>1.988829</td>\n",
       "      <td>3.016109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>2.000000</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>41.000000</td>\n",
       "      <td>45.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ag1k_cluster  multisp_cluster\n",
       "count   1951.000000      1353.000000\n",
       "mean       1.667350         2.254989\n",
       "std        1.988829         3.016109\n",
       "min        1.000000         1.000000\n",
       "25%        1.000000         1.000000\n",
       "50%        1.000000         1.000000\n",
       "75%        2.000000         3.000000\n",
       "max       41.000000        45.000000"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# species per cluster - indication of merged multispecies clusters \n",
    "species_merged_cl = pd.DataFrame({cl:hq_seq_data.groupby([cl,'target'])['species'].nunique().describe()\n",
    "                                  for cl in clustering_types})\n",
    "species_merged_cl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ag1k_cluster</th>\n",
       "      <th>multisp_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>2826.000000</td>\n",
       "      <td>2826.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1.151097</td>\n",
       "      <td>1.079618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.476893</td>\n",
       "      <td>0.284767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>6.000000</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ag1k_cluster  multisp_cluster\n",
       "count   2826.000000      2826.000000\n",
       "mean       1.151097         1.079618\n",
       "std        0.476893         0.284767\n",
       "min        1.000000         1.000000\n",
       "25%        1.000000         1.000000\n",
       "50%        1.000000         1.000000\n",
       "75%        1.000000         1.000000\n",
       "max        6.000000         3.000000"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clusters per species - indications of split species cluster\n",
    "species_split_cl = pd.DataFrame({cl:hq_seq_data.groupby(['species','target'])[cl].nunique().describe()\n",
    "                        for cl in clustering_types}) \n",
    "species_split_cl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ag1k_cluster</th>\n",
       "      <th>multisp_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>7224.000000</td>\n",
       "      <td>7224.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>1.043328</td>\n",
       "      <td>1.029347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.206310</td>\n",
       "      <td>0.170420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>3.000000</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ag1k_cluster  multisp_cluster\n",
       "count   7224.000000      7224.000000\n",
       "mean       1.043328         1.029347\n",
       "std        0.206310         0.170420\n",
       "min        1.000000         1.000000\n",
       "25%        1.000000         1.000000\n",
       "50%        1.000000         1.000000\n",
       "75%        1.000000         1.000000\n",
       "max        3.000000         3.000000"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# same sample genotypes split between clusters - indication of excessive splitting of clusters or contamination \n",
    "sample_split_cl = pd.DataFrame({cl:hq_seq_data.groupby(['s_Sample','target'])[cl].nunique().describe()\n",
    "                                for cl in clustering_types}) \n",
    "sample_split_cl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAEGCAYAAAB/+QKOAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nO3deZgU5bn+8e8NjIKCIIpGQQQTlyAMoOASDKIYcRfEGBVj3EBPjHrMSSKaxHBiLmMiicrRSIhR4oI77saFxAEXFEFRVDQxggLqD0RUVFCW5/dH1YzN0DNTA9PdA3N/rquv6aqufuup7ul+uuqtel5FBGZmZtU1K3UAZmbWODlBmJlZXk4QZmaWlxOEmZnl5QRhZmZ5tSh1APW19dZbR5cuXUodhpnZBmXGjBkfRESH+jxng0sQXbp0Yfr06aUOw8xsgyLp7fo+x4eYzMwsLycIMzPLywnCzMzy2uD6IMys8FasWMH8+fNZvnx5qUOxemrZsiWdOnWirKxsvdtygjCztcyfP582bdrQpUsXJJU6HMsoIli8eDHz58+na9eu692eDzGZ2VqWL1/OVltt5eSwgZHEVltt1WB7fk4QZpaXk8OGqSHfNycIMzPLywnCGqUBAwYwYMCAUodh1qQ5QZjZerv3xQX0u+yfdB35EP0u+yf3vrigqOufMmUKe+yxBy1atOCuu+6qml9RUcERRxyxXm136dKFDz74oN7Pq6io4JlnnlmvdZeaE4SZrZd7X1zAhRNnseCjZQSw4KNlXDhxVlGTROfOnRk/fjwnnnhi0dZZl3VJECtXrixQNOumYAlC0vWSFkp6pZZlBkiaKelVSZMLFYuZFc7lj77BshWr1pi3bMUqLn/0jfVqd/Dgwey5557svvvujBs3DoC//vWv7LLLLgwYMIDhw4fzox/9CEh+5ZeXl9OsWc1fac8//zy9e/fmrbfeyvv4p59+yqmnnkqPHj0oLy/n7rvvXuPxuXPn0r1796rp0aNHM2rUKADGjBlDt27dKC8v5/jjj2fu3LmMHTuWK664gl69evHkk0+yaNEihg4dSt++fenbty9PP/00AKNGjWLEiBEcfPDBnHzyyev8ehVCIa+DGA9cDdyY70FJ7YA/AYdExDuStilgLGZWIO9+tKxe87O6/vrrad++PcuWLaNv374cfvjhXHLJJbzwwgu0adOGAw88kJ49e2Zq65lnnuGcc87hvvvuo3PnznmXueSSS2jbti2zZs0CYMmSJZljveyyy5gzZw6bbropH330Ee3ateOss86idevW/OQnPwHgxBNP5Pzzz2e//fbjnXfeYdCgQcyePRuAGTNm8NRTT9GqVavM6yyGgiWIiJgiqUsti5wITIyId9LlFxYqFjMrnO3btWJBnmSwfbv1+7IbM2YM99xzDwDz5s3jpptuYv/996d9+/YAfPe73+Vf//pXne3Mnj2bESNG8Nhjj7H99tvXuNykSZO47bbbqqa33HLLzLGWl5czbNgwBg8ezODBg2ts/7XXXqua/uSTT1i6dCkARx11VKNLDlDaPohdgC0lVUiaIanGfStJIyRNlzR90aJFRQzRzOry00G70qqs+RrzWpU156eDdl3nNisqKpg0aRJTp07lpZdeonfv3uy667q1t91229GyZUtefPHFWpeLiFqvIWjRogWrV6+ums69GO2hhx7i7LPPZsaMGey55555+xJWr17N1KlTmTlzJjNnzmTBggW0adMGgM0337y+m1UUpUwQLYA9gcOBQcAvJe2Sb8GIGBcRfSKiT4cO9RrvwswKbHDvjvz2mB50bNcKAR3bteK3x/RgcO+O69zmxx9/zJZbbslmm23G66+/zrPPPsvnn3/O5MmTWbJkCStXrlyrj6Am7dq146GHHuKiiy6ioqKixuUOPvhgrr766qrp6oeYtt12WxYuXMjixYv54osvePDBB4Hki3/evHkccMAB/P73v+ejjz7i008/pU2bNlV7CPnanzlzZqb4S6mUCWI+8EhEfBYRHwBTgGwHFM2sURncuyNPjzyQOZcdztMjD1yv5ABwyCGHsHLlSsrLy/nlL3/JPvvsQ8eOHbnooovYe++9Oeigg+jWrRtt27YFkg7oTp06ceedd3LmmWey++67r9HetttuywMPPMDZZ5/Nc889l3edv/jFL1iyZAndu3enZ8+ePPHEE2s8XlZWxsUXX8zee+/NEUccwW677QbAqlWrOOmkk+jRowe9e/fm/PPPp127dhx55JHcc889VZ3UY8aMYfr06ZSXl9OtWzfGjh27Xq9RMSgiCtd40gfxYER0z/PYN0k6sQcBmwDTgOMjosazngD69OkTHlFu41d5kVxtv/iscGbPns03v/nNUoexlk8//ZTWrVuzcuVKhgwZwmmnncaQIUNKHVajk+/9kzQjIvrUp52CdVJLuhUYAGwtaT7wK6AMICLGRsRsSY8ALwOrgevqSg5m1rSNGjWKSZMmsXz5cg4++OAaO4StYRTyLKYTMixzOXB5oWIws43L6NGj17uNG264gauuumqNef369eOaa65Z77Y3Nh4PwsyalFNPPZVTTz211GFsEOpMEOkFbP2A7YFlwCvA9IhYXesTzcxsg1ZjgpB0ADASaA+8CCwEWgKDga9Lugv4Q0R8UoxAzcysuGrbgzgMGF55pXMuSS2AI4DvANlORjYzsw1KjQkiIn4KIKlrRMyp9vAOEXFvQSMzM7OSynKhXL49hLvyzDOzpurlO+CK7jCqXfL35TtKHRGtW7cGkiqsEyZMqJo/ffp0zj333IKue/z48VWVZuvr0ksvbeBo1l2NCULSbpKGAm0lHZNzO4WkL8LMLEkGD5wLH88DIvn7wLmNIknA2gmiT58+jBkzpoQR1W5dEsSqVavqXmgd1LYHsStJP0M74Mic2x7A8IJEY2Ybnn/8GlZUq+a6Ylkyfz3MnTuX3XbbjTPOOIPu3bszbNgwJk2aRL9+/dh5552ZNm0ao0aNWuPaiO7duzN37tw12hk5ciRPPvkkvXr14oorrlhjlLnJkyfTq1cvevXqRe/evVm6dCkVFRX079+fIUOG0K1bN84666w1ivRV98gjj7DHHnvQs2dPBg4cuNbjp5xyyhqj3FXu2bz33nv079+fXr160b17d5588klGjhzJsmXL6NWrF8OGDQPg5ptvZq+99qJXr16ceeaZVcmgdevWVaU/pk6dum4vch1q64O4D7hP0r4RUZi1m9mG7+P59ZtfD2+++SZ33nkn48aNo2/fvkyYMIGnnnqK+++/n0svvZRevXrV2cZll13G6NGjq4rr5ZZvGT16NNdccw39+vXj008/pWXL5ODItGnTeO2119hxxx055JBDmDhxIscee+xabS9atIjhw4czZcoUunbtyocffph52yZMmMCgQYP4+c9/zqpVq/j888/59re/zdVXX11VyG/27NncfvvtPP3005SVlfHDH/6QW265hZNPPpnPPvuM7t278+tfr18irk1tp7n+LCJ+D5woaa2roiOisAfxzGzD0LZTengpz/z11LVrV3r06AHA7rvvzsCBA5FEjx49mDt3bqYEUZt+/frx4x//mGHDhnHMMcfQqVMS81577cVOO+0EwAknnMBTTz2VN0E8++yz9O/fn65duwJUjVWRRd++fTnttNNYsWIFgwcPzrst//jHP5gxYwZ9+/YFYNmyZWyzTTK2WvPmzRk6dGj9NrieajvENDv9Ox2YkedmZgYDL4ayaoPdlLVK5q+nTTfdtOp+s2bNqqabNWvGypUrax2jIYuRI0dy3XXXsWzZMvbZZx9ef/11gLXGhahpnIi6xpCANceRiAi+/PJLAPr378+UKVPo2LEj3//+97nxxrUH34wIfvCDH1SNIfHGG29UDXPasmVLmjdvvtZzGlKNCSIiHkj//i3fraBRmdmGo/w4OHIMtN0BUPL3yDHJ/ALr0qULL7zwAgAvvPACc+ZUPyOftcZlyPWf//yHHj16cMEFF9CnT5+qBDFt2jTmzJnD6tWruf3229lvv/3yPn/fffdl8uTJVevNd4ipS5cuzJiR/Ka+7777WLFiBQBvv/0222yzDcOHD+f000+v2o6ysrKqZQYOHMhdd93FwoULq9p/++23s704DaC2Q0wPADXWAo+IowoSkZlteMqPK0pCqG7o0KHceOON9OrVi759+7LLLmuPOVZeXk6LFi3o2bMnp5xyCr1796567Morr+SJJ56gefPmdOvWjUMPPZSpU6ey7777MnLkSGbNmlXVYZ1Phw4dGDduHMcccwyrV69mm2224fHHH19jmeHDh3P00Uez1157MXDgwKrR4yoqKrj88sspKyujdevWVXsQI0aMoLy8nD322INbbrmF3/zmNxx88MGsXr2asrIyrrnmGnbccceGeglrVeN4EJL2r+2JETG5IBHVweNBNA0eD6K0Gut4EMVQUVGxRqf2hqjg40HkJgBJmwC7kexRvBERX9YvXDMz29BkqeZ6ODAW+A8goKukMyPi74UOzsys2AYMGFC1B5tr77335osvvlhj3k033VR1ltXGKMt4EH8ADoiINwEkfR14CHCCMNuIZTlDpympaSzrxqYhh5HOUotpYWVySL1FUvrbzDZSLVu2ZPHixQ36ZWOFFxEsXry46oK/9ZVlD+JVSQ8Dd5D0QXwXeF7SMWlAExskEjNrNDp16sT8+fNZtGhRqUOxemrZsmXVBX/rK0uCaAn8P6DyrKZFJIMIHUmSMJwgzDYyZWVlVVcHW9NVZ4KICA/eambWBNXZByHp95K2kFQm6R+SPpB0UjGCMzOz0snSSX1wOu70EcB8YBfgpwWNyszMSi5LgihL/x4G3BoR2evZmpnZBitLJ/UDkl4HlgE/lNQBqF/JRDMz2+DUuQcRESOBfYE+EbEC+Bw4utCBmZlZadU2JnVVfduIWBIRq9L7n0XE+2nHdfdiBGlmZsVX2yGmoZJ+DzxCMkDQIpJrIr4BHADsCPxPwSM0M7OSqK2a6/mStgSOJbl6ejuSfojZwJ8j4qnihGhmZqVQayd1RCwB/pLezMysCclymquZmTVBThBmZpZXwRKEpOslLZT0Sh3L9ZW0StKxhYrFzMzqL0stpu9KapPe/4WkiZL2yND2eOCQOtpuDvwOeDRDe2ZmVkRZ9iB+GRFL0+siBgF/A66t60kRMQWoqyzHOcDdeAAiM7NGJ0uCWJX+PRy4NiLuAzZZ3xVL6ggMIRnvuq5lR0iaLmm6BzAxMyuOLAligaQ/A8cBD0vaNOPz6nIlcEHlFdq1iYhxEdEnIvp06NChAVZtZmZ1yVKs7ziSvoTREfGRpO1omHLffYDb0kHRtwYOk7QyIu5tgLbNzGw9ZRlR7nNJC4H9gH8DK9O/6yUiqsYzlDQeeNDJwcys8agzQUj6Fcmv/V2BG0jGh7gZ6FfH824FBgBbS5oP/Cp9LhFRZ7+DmZmVVpZDTEOA3sALABHxbuVpr7WJiBOyBhERp2Rd1szMiiNLZ/OXERFAAEjavLAhmZlZY5AlQdyRnsXUTtJwYBIu3mdmttHL0kk9WtJ3gE9I+iEujojHCx6ZmVmOAQMGAFBRUVHSOJqSLH0QpAnBScHMrAmpMUFIeioi9pO0lLT/ofIhICJii4JHZ2ZmJVPbiHL7pX/rPGPJzMw2Plmque6Te1qrpNaS9i5sWGZmVmpZzmK6Fvg0Z/pzMlRzNTOzDVuWBKH0OggAImI1GTu3zcxsw5UlQbwl6VxJZentPOCtQgdmZmallSVBnAV8C1gAzAf2BkYUMigzMyu9LBfKLQSOL0IsZmbWiGQ5i2kXSf+Q9Eo6XS7pF4UPzczMSinLIaa/ABcCKwAi4mW8R2FmttHLkiA2i4hp1eatLEQwZmbWeGRJEB9I+jpflfs+FnivoFGZmVnJZbme4WxgHLCbpAXAHGBYQaMyM7OSy3IW01vAQelAQc0iYmnhwzIzs1LLchbTVpLGAE8CFZKukrRV4UMzM7NSytIHcRuwCBgKHJvev72QQZmZWell6YNoHxGX5Ez/RtLgQgVkZmaNQ5Y9iCckHS+pWXo7Dnio0IGZmVlpZUkQZwITgC+AL0kOOf1Y0lJJnxQyODMzK50sZzF5RDkzsyYoy1lM/dJTXJF0kqQ/Supc+NDMzKyUso4o97mknsDPgLeBmwoalZmZlVyWBLEyHVHuaOCqiLgK8GEnM7ONXJbTXJdKuhA4CegvqTlQVtiwzMys1LLsQXyP5Aym0yPifaAjcHlBozIzs5LLchbT+8Afc6bfAW4sZFBmZlZ6WfYgzMysCXKCMDOzvGpNEJKaS7q5WMGYmVnjUWuCiIhVQAdJm9S3YUnXS1oo6ZUaHh8m6eX09kx6nYWZmTUSWU5znQs8Lel+4LPKmRHxxxqfkRgPXE3NHdpzgP0jYomkQ0lGrds7QzxmZlYEWRLEu+mtGfW4QC4ipkjqUsvjz+RMPgt0ytq2mZkVXpbTXP8XQNLmEfFZXcuvo9OBv9f0oKQRwAiAzp1dBsrMrBiyFOvbV9JrwOx0uqekPzVUAJIOIEkQF9S0TESMi4g+EdGnQ4cODbVqMzOrRZbTXK8EBgGLASLiJaB/Q6xcUjlwHXB0RCxuiDbNzKxhZLoOIiLmVZu1an1XnJYMnwh8PyL+tb7tmZlZw8rSST1P0reASE93PZf0cFNtJN0KDAC2ljQf+BVpkb+IGAtcDGwF/EkSJFVj+6zLRpiZWcPLkiDOAq4iKdK3AHgUOLuuJ0XECXU8fgZwRob1m5lZCWQ5i+kDYFgRYjEzs0Yky1lMO0l6QNKi9Mro+yTtVIzgrGm698UFzOs9grl7/4R+l/2Te19cUOqQzJqkLJ3UE4A7gO2A7YE7gVsLGZQ1Xfe+uIALJ85i1aZtQWLBR8u4cOIsJwmzEsjSB6GIyB2D+mZJPypUQIU0YMCAUodgdZjXe0SSHHIsW7GK/7lxMleeP65EUVkWFRUVpQ7BGliWBPGEpJHAbUCQjDD3kKT2ABHxYQHjsyZm1SZb1Gu+mRVOlgTxvfTvmdXmn0aSMDaY/gj/wmn8+l32TxZ8tGyt+R233Mzvn1mRZTmLqWsxAjED+OmgXblw4iyWrfjqWsxWZc356aBdSxiVWdPkEeWsURncuyO/PaYHzb/4GCLo2K4Vvz2mB4N7dyx1aGZNTpZDTGZFNbh3x6oOaR9WMisd70GYmVleWS6U6ydp8/T+SZL+KGnHwodmZmallGUP4lrg83TM6J8Bb1PzMKJmZraRyJIgVkZEAEcDV0XEVdRj6FEzM9swZemkXirpQuAkoL+k5qRlu83MbOOVZQ/ie8AXwOkR8T5J2e/LCxqVmZmVXK17EOnews0RcVDlvIh4B/dBmJlt9Grdg4iIVSQd1G1rW87MrKBevoPb9nmFf+7/IlzRHV6+o9QRNQlZ+iCWA7MkPQ58VjkzIs4tWFRmZpVevgMeOJevtVyRTH88Dx5Iv37KjytdXE1AlgTxUHozsxJqquXqb9vnla+SQ6UVy3h/wg85/tw/lSaoEil2ZYEsxfr+JqkV0Dki3ihCTGZmVbbZdEW95lvDqTNBSDoSGA1sAnSV1Av4dUQcVejgzOwrTbYu1RXdk8NK1TRrt0PTfU2KJMtprqOAvYCPACJiJuAS4GZWHAMvhrJWa84ra5XMt4LKeiX1x9XmRSGCMTNbS/lxcOQY3l9exuoA2u4AR45xB3URZOmkfkXSiUBzSTsD5wLPFDYsM7Mc5cdVdUj7sFLxZNmDOAfYneRq6gnAx8B5hQzKzMxKL8sexOER8XPg55UzJH0XuLNgUZmZWcll2YO4MOM8MzPbiNS4ByHpUOAwoKOkMTkPbQGsLHRgZmZWWrUdYnoXmA4cBczImb8UOL+QQZmZWenVmCAi4iXgJUkTImIFgKQtgR0iYkmxAjQzs9LI0gfxuKQtJLUHXgJukPTHAsdlZmYlliVBtI2IT4BjgBsiYk/goDqeY2ZmG7gsCaKFpO2A44AHszYs6XpJCyW9UsPjkjRG0puSXpa0R9a2zcys8LIkiF8DjwJvRsTzknYC/p3heeOBQ2p5/FBg5/Q2Arg2Q5tmZlYkWcp930nORXER8RYwNMPzpkjqUssiRwM3RkQAz0pqJ2m7iHivzqjNzKzgspT7voE8xfki4rT1XHdHILeG7/x03loJQtIIkr0MOnfuvJ6rNTOzLLKU2sjtd2gJDCG5RmJ9Kc+8vFViI2IcMA6gT58+riRrZlYEWQ4x3Z07LelWYFIDrHs+sEPOdCcaJvGYmVkDyNJJXd3OQEMc57kfODk9m2kf4GP3P5iZNR5Z+iCWkhz6Ufr3feCCDM+7FRgAbC1pPvAroAwgIsYCD5PUenoT+Bw4dZ22wMzMCiLLIaY269JwRJxQx+MBnL0ubZuZWeHVVs211gvXIuKFhg/HzMwai9r2IP5Qy2MBHNjAsZiZWSNSWzXXA4oZiJmZNS51nsUk6WxJ7XKmt5T0w8KGZWZmpZblNNfhEfFR5UQ6FsTwwoVkZmaNQZYE0UxS1VXPkpoDmxQuJDMzawyylNp4FLhD0liSzumzgEcKGpWZmZVclgRxAUmhvP8iuVjuMeC6QgZlZmall+VCudXA2PRmZmZNxLrUYjIzsybACcLMzPLKnCAkbV7IQMzMrHHJcqHctyS9BsxOp3tK+lPBIzMzs5LKsgdxBTAIWAwQES8B/QsZlJmZlV6mQ0wRMa/arFUFiMXMzBqRLNdBzJP0LSAkbQKcS3q4yczMNl5Z9iDOIhnYpyPJONK98EA/ZmYbvSwXyn0ADCtCLGZm1ohkGZN6TJ7ZHwPTI+K+hg/JzMwagyyHmFqSHFb6d3orB9oDp0u6soCxmZlZCWXppP4GcGBErASQdC1Jwb7vALMKGJuZmZVQlj2IjkDuVdSbA9tHxCrgi4JEZWZmJZdlD+L3wExJFSTlvvsDl6alNyYVMDYzMyuhLGcx/VXSw8BeJAniooh4N334p4UMzszMSidrsb7lwHvAh8A3JLnUhpnZRi7Laa5nAOcBnYCZwD7AVODAwoZmZmallGUP4jygL/B2RBwA9AYWFTQqMzMruSwJYnlELAeQtGlEvA7sWtiwzMys1LKcxTRfUjvgXuBxSUuAd+t4jpmZbeCynMU0JL07StITQFvgkYJGZWZmJVdrgpDUDHg5IroDRMTkokRlZmYlV2sfRESsBl6S1LlI8ZiZWSORpQ9iO+BVSdOAzypnRsRRBYvKzMxKLkuC+N91bVzSIcBVQHPguoi4rNrjbYGbgc5pLKMj4oZ1XZ+ZmTWcLJ3UkyXtCOwcEZMkbUbyhV8rSc2Ba0iqvs4Hnpd0f0S8lrPY2cBrEXGkpA7AG5JuiYgv12lrzMyswdR5HYSk4cBdwJ/TWR1JTnmty17AmxHxVvqFfxtwdLVlAmgjSUBrklIeKzPGbmZmBZTlQrmzgX7AJwAR8W9gmwzP6wjMy5men87LdTXwTZLrKmYB56Ud42uQNELSdEnTFy3yRdxmZsWQpQ/ii4j4MvmRD5JakPzyr4vyzKv+vEEk9Z0OBL5OciHekxHxyRpPihgHjAPo06dPlnWb2UamoqKi1CE0OVn2ICZLughoJek7wJ3AAxmeNx/YIWe6E2tfgX0qMDESbwJzgN0ytG1mZgWWJUGMJCnONws4E3gY+EWG5z0P7Cypq6RNgOOB+6st8w4wEEDStiQ1nt7KFrqZmRVSlkNMRwM3RsRf6tNwRKyU9CPgUZKznq6PiFclnZU+Pha4BBgvaRbJIakLIuKDem2BmZkVRJYEcRRwpaQpJGciPRoRmc40ioiHSfY4cueNzbn/LnBw9nDNzKxY6jzEFBGnAt8g6Xs4EfiPpOsKHZiZmZVWlj0IImKFpL+TnIXUiuSw0xmFDMzMzEory4Vyh0gaD7wJHAtcR1KfyczMNmJZ9iBOIel7ODMivihsOGZm1lhkqcV0fO60pH7AiRFxdsGiMjOzksvUByGpF0kH9XEkF7NNLGRQZmZWejUmCEm7kFzcdgKwGLgdUEQcUKTYzMyshGrbg3gdeBI4Mi2DgaTzixKVmZmVXG1nMQ0F3geekPQXSQPJX4DPzMw2QjUmiIi4JyK+R1I8rwI4H9hW0rWSfPWzmdlGLsuV1J9FxC0RcQRJRdaZJAX8zMxsI5almmuViPgwIv4cEQcWKiAzM2sc6pUgzMys6XCCMDOzvJwgzMwsLycIMzPLywnCzMzycoIwM7O8nCDMzCwvJwgzM8srU7lvs2KrqKgodQhmTZ73IMzMLC8nCDMzy8sJwszM8nKCMDOzvJwgzMwsLycIMzPLywnCzMzycoIwM7O8nCDMzCwvRUSpY6gXSYuAtxugqa2BDxqgnWJyzMXhmIvDMRdHZcw7RkSH+jxxg0sQDUXS9IjoU+o46sMxF4djLg7HXBzrE7MPMZmZWV5OEGZmlldTThDjSh3AOnDMxeGYi8MxF8c6x9xk+yDMzKx2TXkPwszMauEEYWZmeW3UCULS9ZIWSnqlhseHSXo5vT0jqWexY8wTU60x5yzXV9IqSccWK7ZaYqkzZkkDJM2U9KqkycWMr4Z46vrfaCvpAUkvpTGfWuwY88S0g6QnJM1OYzovzzKSNEbSm+n/9R6liDUnniwxN6rPYZaYc5ZtFJ/DrDHX+3MYERvtDegP7AG8UsPj3wK2TO8fCjzX2GNOl2kO/BN4GDi2sccMtANeAzqn09tsADFfBPwuvd8B+BDYpMQxbwfskd5vA/wL6FZtmcOAvwMC9in1/3TGmBvV5zBLzOljjeZzmPF1rvfncKPeg4iIKSQf7JoefyYilqSTzwKdihJYLeqKOXUOcDewsPAR1S1DzCcCEyPinXT5ksedIeYA2kgS0DpddmUxYqsxoIj3IuKF9P5SYDbQsdpiRwM3RuJZoJ2k7YocapUsMTe2z2HG1xka0ecwY8z1/hxu1Amink4n+eXVqEnqCAwBxpY6lnrYBdhSUoWkGZJOLnVAGVwNfBN4F5gFnBcRq0sb0lckdQF6A89Ve6gjMC9nej75v9yKrpaYczWqz2FNMTfmz2Etr3O9P4ctGj68DY+kA0j+MfcrdSwZXAlcEBGrkh+3G4QWwJ7AQKAVMFXSs9VaQ+gAAAh4SURBVBHxr9KGVatBwEzgQODrwOOSnoyIT0obFkhqTfLL9b/zxJPvn6Lk57LXEXPlMo3qc1hHzI3yc1hHzPX+HDb5BCGpHLgOODQiFpc6ngz6ALel/5RbA4dJWhkR95Y2rFrNBz6IiM+AzyRNAXqSHCdtrE4FLovkYO2bkuYAuwHTShmUpDKSL4BbImJinkXmAzvkTHci2QsqmQwxN7rPYYaYG93nMOP/Rr0+h036EJOkzsBE4PuN/NdslYjoGhFdIqILcBfww0aeHADuA74tqYWkzYC9SY6RNmbvkPzSQtK2wK7AW6UMKO0P+SswOyL+WMNi9wMnp2cz7QN8HBHvFS3IarLE3Ng+h1libmyfw4z/G/X+HG7UexCSbgUGAFtLmg/8CigDiIixwMXAVsCf0l8CK6PElRozxNzo1BVzRMyW9AjwMrAauC4iaj2Nt9AyvM6XAOMlzSI5bHNBRJS6zHM/4PvALEkz03kXAZ2hKu6HSc5kehP4nGRPqJSyxNzYPodZYm5s6ox5XT6HLrVhZmZ5NelDTGZmVjMnCDMzy8sJwszM8nKCMDOzvJwgzMwsLyeIJkLSEEkhabdSx1JKkraXdFep48hC0vh1qRIqqYukExtg/ZL0T0lbrG9bDUHSaEkHljqOpsQJouk4AXgKOL4hGpPUvCHaKbaIeDciSl2audDXH3UhKcyWWQ3v52HASw1dXmQ9tv//gJENGYvVzgmiCUjrs/QjqXNzfM782yUdljM9XtJQSc0lXS7p+bRG/5np4wPSmvMTSArYIenetPDXq5JG5LR1uqR/pYXB/iLp6nR+B0l3p20/L6lfnnh3lzRNSd36lyXtnP4qfl3S39J5d6VXgyJpT0mT0zgeVVq9VNI3JE1SMqbDC5K+nrbzSvp4Tdu5naQp6fpfkfTtPDHOlfS7NM5pkr5R2/ZJGiVpnKTHgBvztPczSbPSWC+rYX1bp/f7SKpI7++fxjlT0ouS2gCXkVwxO1PS+fV5P6sZRnL1beVeyeuSrktfk1skHSTpaUn/lrRXutzmSsbaeD6N5+h0/imS7pT0APCYpGaS/pT+3zwo6WGle0s1vZ8R8TawlaSv5YnVCqEUtct9K+4NOAn4a3r/Gb6qGz8E+Ft6fxOSKqCtgBHAL9L5mwLTga4kVx5/BnTNabt9+rcV8ArJFbHbA3OB9iRXJz8JXJ0uNwHYL73fmaQ0QPV4/w8YlhNXK5JfxQH0S+dfD/wkbf8ZoEM6/3vA9en954Ah6f2WwGZpO6+k82razv8Bfp7Obw60yRPj3JxlTgYerG37gFHADKBVnrYOTbdhs2qv6XjScQbS9W2d3u8DVKT3H8h5TVqTVEcYUBlPHdu51vtZLa63K7c9fd1WAj1IfljOSN8DkZQYvzdd7lLgpPR+O5I6P5sDp5DUAqrctmNJrvpuBnwNWJLOq/H9TKf/Agwt9Weqqdw26lIbVuUEkuqTALel0y+QlFUeI2lT4BBgSkQsk3QwUK6vjn+3BXYGvgSmRcScnLbPlTQkvb9DutzXgMkR8SGApDtJSg0DHAR001cVMLeQ1CaSGvaVpgI/l9SJpH79v9Pl50XE0+kyNwPnAo8A3UmqrULyhf5e+ku6Y0TcAxARy9NYcl+XmrbzeeB6JcXP7o2ImeR3a87fK2rbvvT+/RGxLE87BwE3RMTnaax1jQeS62ngj5JuIXmt5mvt6qL1eT9zta/2vsyJiMo9x1eBf0REKClH0iVnXUdJ+kk63ZK03APweM627QfcGUkJ9fclPZHO35U872dODAtJfoBYEThBbOQkbUVSsrq7pCD5wIWkn0XE8vRQxSCSX2qVX3gCzomIR6u1NYDkF2fu9EHAvhHxedpWS/KXnK7ULF0+3xclABExQdJzwOHAo5LOICmUV70uTKTrejUi9q0Wa5aO1bzbmT6/f7r+myRdHhFrHRaqFk/l/bzbl37ZfUZ+ou6S3Cv56pBwy6qVRlwm6SGS/oJnJR1UQ/t1vp/51impWXw1DsYXOY+tzplezVffJSL5hf9GtXXtXW1dNf2P5H0/c7QEavzfsYblPoiN37EkI4ztGEn1yR2AOXxVc/82koJu3wYqv0AeBf4r/QWNpF0kbZ6n7bbAkjQ57EYyxCUkJbH3l7Slkg7JoTnPeQz4UeWEpF7VG5W0E/BWRIwhqU5anj7UWVLlF0dlp/sbQIfK+ZLKJO0eScfqfEmD0/mbKu2zyJF3OyXtCCyMiL+QVMisaVzn7+X8nZp1+/J4DDhNX/WptM+zzFySWv6Q83pK+npEzIqI35EcOtoNWEoy7GSt25khrjeAnTIsl+tR4BylGVFS7xqWewoYmvZFbEtyuKtynWu9nznP24XkUKYVgRPExu8E4J5q8+7mq7NcHiMZn3lSRHyZzruOZOzaF5R06P6Z/HubjwAtJL1MUv30WYCIWEByLPo5YFLa1sfpc84F+qSdpa8BZ+Vp93vAK0qqUu7GV526s4EfpOtrD1ybxnws8DtJL5EM8vOtdPnvkxwCe5nkuHb1zs2atnMAMFPSiyRfxlfliRFg03RP5zzg/Hps3xoi4hGSRDg93eaf5Fnsf4GrJD0JrMqZ/99pp/FLJL+s/05SrXOlkg7v82vZzro8xFdf3FldQtKP8HK6rktqWO5ukj6JynieIylNXuP7mSa4b5AkQisCV3O1gpDUOiI+Tfcg7iHpaKyeqOrTXheSjtfuDRTiepE0F+gTpS8BXjDp2UM3RsR3CtR+5f/IViR7nf0i4v1alh9CcoLFLwsRj63NfRBWKKPS4+EtSfZSGvugRlZNRLyn5BTlLaIwQ60+KKkdyZlql9SWHFItgD8UIA6rgfcgzMwsL/dBmJlZXk4QZmaWlxOEmZnl5QRhZmZ5OUGYmVle/x9ftkUy874k2wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots()\n",
    "ax.set_xlabel('Average species per cluster (merge)')\n",
    "ax.set_ylabel('Average clusters per species (split)')\n",
    "for cl in clustering_types:\n",
    "    mean_merge = species_merged_cl.loc['mean',cl]\n",
    "    std_merge = species_merged_cl.loc['std',cl]\n",
    "    mean_split = species_split_cl.loc['mean',cl]\n",
    "    std_split = species_split_cl.loc['std',cl]\n",
    "    ax.plot(mean_merge,\n",
    "            mean_split, 'o', label=cl)\n",
    "    ax.vlines(mean_merge, mean_split-std_split, mean_split+std_split)\n",
    "    ax.hlines(mean_split, mean_merge-std_split, mean_merge+std_split)\n",
    "    ax.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clustering with multi-species thresholds results in significant increase of merging several species in one cluster compared to ag1000g derived thresholds. On the other hand, decrease of split of species between clusters is quite moderate, thus use of ag1000g thresholds would give a more fine-grained picture without much false splits. In principle, the difference could be more pronounced for amplicons where difference between ag1k and multisp thresholds is the biggest."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set clustering column to be used for filtering\n",
    "cl_col = 'ag1k_cluster'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Outlier samples within species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target</th>\n",
       "      <th>ag1k_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>328.000000</td>\n",
       "      <td>328.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>31.716463</td>\n",
       "      <td>2.301829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>16.092112</td>\n",
       "      <td>0.679807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>18.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>32.000000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>45.250000</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>61.000000</td>\n",
       "      <td>6.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           target  ag1k_cluster\n",
       "count  328.000000    328.000000\n",
       "mean    31.716463      2.301829\n",
       "std     16.092112      0.679807\n",
       "min      0.000000      2.000000\n",
       "25%     18.000000      2.000000\n",
       "50%     32.000000      2.000000\n",
       "75%     45.250000      2.000000\n",
       "max     61.000000      6.000000"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find cases where genotypes of a single species are split between clusters \n",
    "cl_per_spt = hq_seq_data.groupby(['species','target'])[cl_col].nunique()\n",
    "multicl_per_spt = cl_per_spt[cl_per_spt > 1].reset_index()\n",
    "multicl_per_spt.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "species\n",
       "Anopheles_hyrcanus       34.0\n",
       "Anopheles_nili           22.0\n",
       "Anopheles_tenebrosus     20.0\n",
       "Anopheles_theileri       19.0\n",
       "Anopheles_rampae         18.0\n",
       "Anopheles_coustani       17.0\n",
       "Anopheles_ziemanni       17.0\n",
       "Anopheles_gambiae        16.0\n",
       "Anopheles_coluzzii       16.0\n",
       "Anopheles_tessellatus    13.0\n",
       "Name: ag1k_cluster, dtype: float64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# species with split genotypes\n",
    "(cl_per_spt > 1).sum(level='species') \\\n",
    "          .sort_values(ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "target\n",
       "50    16.0\n",
       "19    14.0\n",
       "47    14.0\n",
       "40    13.0\n",
       "16    12.0\n",
       "30    12.0\n",
       "32    12.0\n",
       "41    12.0\n",
       "43    12.0\n",
       "18    12.0\n",
       "Name: ag1k_cluster, dtype: float64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# targets with split genotypes\n",
    "(cl_per_spt > 1).sum(level='target') \\\n",
    "          .sort_values(ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Top targets also have significant number of within-species distances higher than ag1k threshold (see 3_distances notebook)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(292,)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "species             target  ag1k_cluster\n",
       "Anopheles_aconitus  29      cl-29-13        VBS00055\n",
       "                            cl-29-4         VBS00053\n",
       "                    41      cl-41-15        VBS00053\n",
       "                            cl-41-21        VBS00055\n",
       "                    43      cl-43-2         VBS00053\n",
       "Name: s_Sample, dtype: object"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# clusters containing only a single sample from the species\n",
    "# TODO - double-check that this actually works as expected\n",
    "outlier_gt = []\n",
    "for _, r in multicl_per_spt.iterrows():\n",
    "    # subset sequencing data to problematic species and target\n",
    "    sub_data = hq_seq_data[(hq_seq_data.species == r.species) & \n",
    "                         (hq_seq_data.target == r.target)]\n",
    "    # count samples per cluster\n",
    "    sp_group = sub_data.groupby(['species','target',cl_col])['s_Sample'].unique()\n",
    "    # one off - one sample per cluster for a given species-target\n",
    "    outlier_gt.append(sp_group[sp_group.apply(lambda x: len(x)) == 1])\n",
    "# convert to series\n",
    "outlier_gt = pd.concat(outlier_gt).apply(lambda x: x[0])\n",
    "display(outlier_gt.shape)\n",
    "outlier_gt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Azie-334                                      13\n",
       "VBS00149                                      13\n",
       "Acol-645                                       9\n",
       "VBS00115                                       9\n",
       "Apal-257                                       7\n",
       "VBS00122                                       7\n",
       "VBS00112                                       7\n",
       "Apal-81                                        7\n",
       "VBS00108                                       7\n",
       "anopheles-sinensis-sinensisscaffoldsasins2     6\n",
       "Name: s_Sample, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# most frequently excluded samples\n",
    "outlier_gt.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(292,)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(273,)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# exclude reference genomes from outliers - those are expected to be ground truth\n",
    "display(outlier_gt.shape)\n",
    "outlier_gt = outlier_gt[~outlier_gt.str.startswith('anopheles')]\n",
    "display(outlier_gt.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "outlier_gt = outlier_gt.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "species\ttarget\tag1k_cluster\ts_Sample\r\n",
      "Anopheles_aconitus\t29\tcl-29-13\tVBS00055\r\n",
      "Anopheles_aconitus\t29\tcl-29-4\tVBS00053\r\n",
      "Anopheles_aconitus\t41\tcl-41-15\tVBS00053\r\n",
      "Anopheles_aconitus\t41\tcl-41-21\tVBS00055\r\n",
      "Anopheles_aconitus\t43\tcl-43-2\tVBS00053\r\n",
      "Anopheles_aconitus\t43\tcl-43-3\tVBS00055\r\n",
      "Anopheles_aconitus\t52\tcl-52-24\tVBS00053\r\n",
      "Anopheles_aconitus\t52\tcl-52-31\tVBS00055\r\n",
      "Anopheles_aconitus\t55\tcl-55-10\tVBS00053\r\n"
     ]
    }
   ],
   "source": [
    "# write file\n",
    "outlier_gt.to_csv(OUTLIER_GENOTYPES, index=False, sep='\\t')\n",
    "! head {OUTLIER_GENOTYPES}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get outlier sequence status by matching original sequences with outliers on 4 cols\n",
    "seq_data = pd.merge(seq_data, \n",
    "                    outlier_gt, \n",
    "                    on=('species','target','ag1k_cluster','s_Sample'), \n",
    "                    indicator=True, \n",
    "                    how='outer')\n",
    "# make sure no unmatched genotypes in outlier exist\n",
    "assert (seq_data['_merge'] == 'right_only').sum() == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert indicator to outlier genotype\n",
    "seq_data = seq_data.rename(columns={'_merge':'outlier_genotype'})\n",
    "seq_data['outlier_genotype'] = seq_data['outlier_genotype'] \\\n",
    "        .replace({'left_only':False,\n",
    "                  'both':True})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter allelic splits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAD4CAYAAAAD6PrjAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAPeklEQVR4nO3df4xldX3G8fcjSASri8pazQIOCkW3jaY4Utv6W9vyIyuitICmppYuVautaWzYGqMmxgSTtlqiFldrqLaFICq6BaQaozQVK4ulAoJ1g1RXmoDSLAWpCHz6x737dVxnmTOz98y5d/b9Sia558w95z45O7PPfM/PVBWSJAE8bOgAkqTpYSlIkhpLQZLUWAqSpMZSkCQ1Bw4dYF8cdthhNTc3N3QMSZop11577ferav1i35vpUpibm2P79u1Dx5CkmZLkv/b2PXcfSZIaS0GS1ExVKSR5WZIPJfl0kt8cOo8k7W96L4UkH0lye5Ib9ph/QpJvJtmRZAtAVV1aVZuB3wNO7zubJOmnrcZI4QLghIUzkhwAvB84EdgInJlk44K3vHX8fUnSKuq9FKrqKuDOPWYfD+yoqluq6j7gIuCUjLwbuKKqvrbY+pKcnWR7ku133HFHv+ElaT8z1DGFDcB3F0zvHM97I/AS4LQkr11swaraWlXzVTW/fv2ip9lKklZoqOsUssi8qqrzgPNWO4wkaWSoUtgJHLFg+nDgttUMMLflshUve+u5J08wiSRNj6F2H10DHJPkqCQHAWcAnxkoiyRpbDVOSb0QuBo4NsnOJGdV1f3AG4ArgZuAi6vqxmWsc1OSrbt27eontCTtp3rffVRVZ+5l/uXA5Stc5zZg2/z8/OZ9ySZJ+mlTdUWzJGlYloIkqbEUJEnNTJaCB5olqR8zWQpVta2qzl63bt3QUSRpTZnJUpAk9cNSkCQ1loIkqbEUJEnNTJaCZx9JUj9mshQ8+0iS+jGTpSBJ6oelIElqLAVJUmMpSJKamSwFzz6SpH7MZCl49pEk9WMmS0GS1A9LQZLUWAqSpMZSkCQ1loIkqbEUJEnNTJaC1ylIUj9mshS8TkGS+jGTpSBJ6oelIElqLAVJUmMpSJIaS0GS1FgKkqTGUpAkNZaCJKmZyVLwimZJ6seBQwdYiaraBmybn5/fPMTnz225bJ+Wv/XckyeURJImayZHCpKkflgKkqTGUpAkNZaCJKmxFCRJjaUgSWosBUlSYylIkhpLQZLUWAqSpMZSkCQ1M1kK3hBPkvoxk6VQVduq6ux169YNHUWS1pSZLAVJUj8sBUlSYylIkhpLQZLUWAqSpMZSkCQ1loIkqbEUJEmNpSBJaiwFSVJjKUiSGktBktRYCpKkxlKQJDWWgiSpmclS8CE7ktSPmSwFH7IjSf2YyVKQJPXDUpAkNZaCJKmxFCRJjaUgSWosBUlS06kUkvxS30EkScPrOlI4P8lXk7w+yaG9JpIkDaZTKVTVc4BXAUcA25P8Y5Lf6DWZJGnVdT6mUFXfAt4KnAM8Hzgvyc1JXt5XOEnS6up6TOHpSd4D3AS8CNhUVU8bv35Pj/kkSavowI7vex/wIeAtVXXv7plVdVuSt/aSTJK06rqWwknAvVX1AECShwGPqKofVtXHeksnSVpVXY8pfB44eMH0IeN5kqQ1pOtI4RFVdffuiaq6O8khPWVa8+a2XLbiZW899+QJJpGkn9Z1pHBPkuN2TyR5JnDvQ7xfkjSDuo4U3gR8PMlt4+knAqf3E0mSNJROpVBV1yR5KnAsEODmqvpxr8m0KHc9SepT15ECwLOAufEyv5yEqvpoL6kkSYPoVApJPgY8BbgOeGA8uwBLQZLWkK4jhXlgY1VVn2EkScPqevbRDcAT+gwiSRpe15HCYcA3knwV+NHumVX10l5SSZIG0bUU3tFnCEnSdOh6SuqXkjwJOKaqPj++mvmAfqNJklZb11tnbwYuAT44nrUBuHSSQZI8OcnfJrlkkuuVJHXX9UDzHwG/DtwF7YE7j19qoSQfSXJ7khv2mH9Ckm8m2ZFky3idt1TVWcuLL0mapK6l8KOqum/3RJIDGV2nsJQLgBMWzkhyAPB+4ERgI3Bmko0dc0iSetS1FL6U5C3AweNnM38c2LbUQlV1FXDnHrOPB3aMRwb3ARcBp3QNnOTsJNuTbL/jjju6LiZJ6qBrKWwB7gCuB/4QuJzR85pXYgPw3QXTO4ENSR6X5HxGt9D4870tXFVbq2q+qubXr1+/wgiSpMV0PfvoQUaP4/zQBD4zi39E/QB47QTWL0laoa73Pvo2ixxDqKonr+AzdwJHLJg+HLhtL++VJK2i5dz7aLdHAL8NPHaFn3kNcEySo4DvAWcAr1zhuiRJE9TpmEJV/WDB1/eq6r3Ai5ZaLsmFwNXAsUl2Jjmrqu4H3gBcCdwEXFxVNy4ndJJNSbbu2rVrOYtJkpbQdffRcQsmH8Zo5PCopZarqjP3Mv9yRgerV6SqtgHb5ufnN690HZKkn9V199FfLnh9P3Ar8DsTTyNJGlTXs49e2HcQSdLwuu4++tOH+n5V/dVk4kiShtT14rV54HWMLjzbwOh6go2MjisseWxh0jzQLEn9WM5Ddo6rqv8FSPIO4ONV9Qd9BXsoHmiWpH50HSkcCdy3YPo+YG7iaSRJg+o6UvgY8NUkn2J0ZfOpwEd7SyVJGkTXs4/eleQK4LnjWa+pqn/vL5YkaQhddx8BHALcVVV/Dewc36ZCkrSGdH0c59uBc4Ddt7R+OPD3fYXqkMezjySpB11HCqcCLwXuAaiq2xjgVNTdqmpbVZ29bt26oSJI0prUtRTuq6pifPvsJI/sL5IkaShdS+HiJB8EDk2yGfg8k3ngjiRpinQ9++gvxs9mvgs4FnhbVX2u12SSpFW3ZCkkOQC4sqpeAlgEkrSGLbn7qKoeAH6YxKO6krTGdb2i+f+A65N8jvEZSABV9ce9pFpCkk3ApqOPPnqIj5ekNatrKVw2/poK3hBPkvrxkKWQ5Miq+k5V/d1qBZIkDWepYwqX7n6R5BM9Z5EkDWypUsiC10/uM4gkaXhLlULt5bUkaQ1a6kDzM5LcxWjEcPD4NePpqqpH95pOkrSqHrIUquqA1QoiSRrecp6nMDW8dbYk9WMmS8FbZ0tSP2ayFCRJ/bAUJEmNpSBJaiwFSVJjKUiSGktBktRYCpKkZiZLwYvXJKkfM1kKXrwmSf2YyVKQJPXDUpAkNZaCJKmxFCRJjaUgSWosBUlSYylIkhpLQZLUWAqSpMZSkCQ1loIkqZnJUvCGeJLUj5ksBW+IJ0n9mMlSkCT1w1KQJDWWgiSpsRQkSY2lIElqLAVJUmMpSJIaS0GS1FgKkqTGUpAkNZaCJKmxFCRJjaUgSWosBUlSYylIkhpLQZLUzGQp+OQ1SerHTJaCT16TpH7MZClIkvphKUiSGktBktRYCpKkxlKQJDWWgiSpsRQkSY2lIElqLAVJUmMpSJIaS0GS1FgKkqTGUpAkNZaCJKmxFCRJjaUgSWosBUlSYylIkhpLQZLUWAqSpMZSkCQ1loIkqbEUJEmNpSBJaiwFSVJz4NABdkvySOADwH3AF6vqHwaOJEn7nV5HCkk+kuT2JDfsMf+EJN9MsiPJlvHslwOXVNVm4KV95pIkLa7v3UcXACcsnJHkAOD9wInARuDMJBuBw4Hvjt/2QM+5JEmL6HX3UVVdlWRuj9nHAzuq6haAJBcBpwA7GRXDdTxEWSU5Gzgb4Mgjj5x86DVsbstlK1721nNPnmASSfvy+wj9/U4OcaB5Az8ZEcCoDDYAnwRekeRvgG17W7iqtlbVfFXNr1+/vt+kkrSfGeJAcxaZV1V1D/Ca1Q4jSfqJIUYKO4EjFkwfDtw2QA5J0h6GKIVrgGOSHJXkIOAM4DMD5JAk7aHvU1IvBK4Gjk2yM8lZVXU/8AbgSuAm4OKqunGZ692UZOuuXbsmH1qS9mN9n3105l7mXw5cvg/r3QZsm5+f37zSdUiSfpa3uZAkNZaCJKmxFCRJTapq6AzLlmQTsAk4HfjWCldzGPD9iYWaHHMtj7mWx1zdTWMmmEyuJ1XVolf/zmQpTEKS7VU1P3SOPZlrecy1PObqbhozQf+53H0kSWosBUlSsz+XwtahA+yFuZbHXMtjru6mMRP0nGu/PaYgSfpZ+/NIQZK0B0tBktSs6VLY2zOiF3w/Sc4bPyv660mOm5JcT01ydZIfJXnzamTqmOtV4+309SRfTvKMKcl1yjjTdUm2J3nONORa8L5nJXkgyWnTkCvJC5LsGm+v65K8bRpyLch2XZIbk3xpGnIl+bMF2+qG8b/lYwfOtC7JtiT/Md5Wk3sWTVWt2S/gecBxwA17+f5JwBWMHvzzbODfpiTX44FnAe8C3jxF2+vXgMeMX584Rdvr5/jJ8bGnAzdPQ67xew4AvsDoBpCnTUMu4AXAP63Wz9Uych0KfAM4cjz9+GnItcd7NwFfGDoT8Bbg3ePX64E7gYMm8dlreqRQVVcx2lh7cwrw0Rr5CnBokicOnauqbq+qa4Af951lj89dKteXq+p/xpNfYfSApGnIdXeNfzuARwKrcvZEh58vgDcCnwBu7z/RSMdcq65DrlcCn6yq74zfvyrbbJnb60zgwh7jAJ0yFfCoJGH0R9GdwP2T+Ow1XQod7O150VraWYxGWVMhyalJbgYuA35/6DwASTYApwLnD51lEb863vVwRZJfHDrM2C8Aj0nyxSTXJnn10IEWSnIIcAKjkh/a+4CnMXpq5fXAn1TVg5NY8RDPaJ4miz4vetVTzJgkL2RUCquy776LqvoU8KkkzwPeCbxk4EgA7wXOqaoHRn/QTY2vMbr3zd1JTgIuBY4ZOBOM/j96JvBi4GDg6iRfqar/HDZWswn416qahlHYbwHXAS8CngJ8Lsm/VNVd+7ri/X2k4POilynJ04EPA6dU1Q+GzrOn8bD7KUkOGzoLMA9clORW4DTgA0leNmwkqKq7quru8evLgYdPyfbaCXy2qu6pqu8DVwGrcjJDR2ewCruOOnoNo11tVVU7gG8DT53Eivf3UvgM8OrxWUjPBnZV1X8PHWpaJTkS+CTwu1P01xtJjh7vW2V8BtlBwOCFVVVHVdVcVc0BlwCvr6pLB45Fkics2F7HM/p/YPDtBXwaeG6SA8e7an6F0SN7B5dkHfB8RhmnwXcYjahI8vPAscAtk1jxmt59lNEzol8AHJZkJ/B24OEAVXU+ozNCTgJ2AD9k1L6D50ryBGA78GjgwSRvAjZOYmi4L7mAtwGPY/QXL8D9tQp3keyQ6xWMyv3HwL3A6QsOPA+ZaxAdcp0GvC7J/Yy21xnTsL2q6qYknwW+DjwIfLiqHvJ039XINX7bqcA/V9U9fefpmOmdwAVJrme0G/yc8ehq3z97FX4WJEkzYn/ffSRJWsBSkCQ1loIkqbEUJEmNpSBJaiwFSVJjKUiSmv8HGXpqJBsutWYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# for all samples\n",
    "cl_per_sample = seq_data.groupby(['s_Sample','target']) \\\n",
    "                             [cl_col].nunique()\n",
    "                       \n",
    "cl_per_sample.mean(level='s_Sample') \\\n",
    "      .sort_values(ascending=False) \\\n",
    "      .plot.hist(bins=20, logy=True);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "split_alleles = cl_per_sample[cl_per_sample > 1].reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(363, 3)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# total number of split allelic genotypes\n",
    "split_alleles.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "VBS00145    40\n",
       "VBS00112     9\n",
       "VBS00149     9\n",
       "Athe-1       8\n",
       "Athe-6-1     8\n",
       "Name: s_Sample, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# most frequently split samples - cross-contamination?\n",
    "split_alleles.s_Sample.value_counts().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "50    25\n",
       "47    18\n",
       "17    17\n",
       "16    17\n",
       "32    17\n",
       "10    16\n",
       "30    15\n",
       "40    15\n",
       "43    13\n",
       "19    12\n",
       "Name: target, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# most frequently split targets - overly specific clustering\n",
    "# amplicons are similar to the ones identified in outlier genotypes\n",
    "split_alleles.target.value_counts().head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "s_Sample\ttarget\tag1k_cluster\r\n",
      "Abro-30\t28\t2\r\n",
      "Acar-191\t16\t2\r\n",
      "Acar-192\t10\t2\r\n",
      "Acar-192\t16\t2\r\n",
      "Acar-192\t19\t2\r\n",
      "Acar-192\t37\t2\r\n",
      "Acar-192\t41\t2\r\n",
      "Acar-192\t55\t2\r\n",
      "Acar-193\t10\t2\r\n"
     ]
    }
   ],
   "source": [
    "# write to file\n",
    "split_alleles.to_csv(SPLIT_ALLELES, index=False, sep='\\t')\n",
    "! head {SPLIT_ALLELES}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_data = pd.merge(seq_data, \n",
    "                    split_alleles[['s_Sample','target']], \n",
    "                    on=['s_Sample','target'], \n",
    "                    indicator=True, \n",
    "                    how='outer')\n",
    "# make sure no unmatched genotypes in outlier exist\n",
    "assert (seq_data['_merge'] == 'right_only').sum() == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert indicator to outlier genotype\n",
    "seq_data = seq_data.rename(columns={'_merge':'split_alleles'})\n",
    "seq_data['split_alleles'] = seq_data['split_alleles'] \\\n",
    "        .replace({'left_only':False,\n",
    "                  'both':True})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Check and write"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s_Sample</th>\n",
       "      <th>target</th>\n",
       "      <th>consensus</th>\n",
       "      <th>reads</th>\n",
       "      <th>species</th>\n",
       "      <th>combUID</th>\n",
       "      <th>ag1k_cluster</th>\n",
       "      <th>multisp_cluster</th>\n",
       "      <th>outlier_genotype</th>\n",
       "      <th>split_alleles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abro-21</td>\n",
       "      <td>13</td>\n",
       "      <td>TCTCCATTTCTATCTGATGAAACTTGTGCCCGTGGTTTAATTAATG...</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Anopheles_brohieri</td>\n",
       "      <td>13-0</td>\n",
       "      <td>cl-13-18</td>\n",
       "      <td>cl-13-11</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Abro-21</td>\n",
       "      <td>17</td>\n",
       "      <td>GGTCTAGACCCGGAAGCCCTCACCGGCAAATACCCGGTAAGTGTTT...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Anopheles_brohieri</td>\n",
       "      <td>17-0</td>\n",
       "      <td>cl-17-8</td>\n",
       "      <td>cl-17-5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Abro-21</td>\n",
       "      <td>18</td>\n",
       "      <td>AAAAAGGATGAGGTGAGTACCAGACTCCTCCCAGAATGATTATAAT...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>Anopheles_brohieri</td>\n",
       "      <td>18-0</td>\n",
       "      <td>cl-18-21</td>\n",
       "      <td>cl-18-17</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Abro-21</td>\n",
       "      <td>19</td>\n",
       "      <td>TATTTGTCCACCTTTTGCCGACCGAGTTTTCTTGCCCGAGACGCGT...</td>\n",
       "      <td>23.0</td>\n",
       "      <td>Anopheles_brohieri</td>\n",
       "      <td>19-0</td>\n",
       "      <td>cl-19-41</td>\n",
       "      <td>cl-19-22</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Abro-21</td>\n",
       "      <td>24</td>\n",
       "      <td>AATGTAGCCCCGAGCGCGAAGCCGTGAAGCTTCTGCGCGCCACCTC...</td>\n",
       "      <td>36.0</td>\n",
       "      <td>Anopheles_brohieri</td>\n",
       "      <td>24-0</td>\n",
       "      <td>cl-24-1</td>\n",
       "      <td>cl-24-1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  s_Sample  target                                          consensus  reads  \\\n",
       "0  Abro-21      13  TCTCCATTTCTATCTGATGAAACTTGTGCCCGTGGTTTAATTAATG...   23.0   \n",
       "1  Abro-21      17  GGTCTAGACCCGGAAGCCCTCACCGGCAAATACCCGGTAAGTGTTT...   14.0   \n",
       "2  Abro-21      18  AAAAAGGATGAGGTGAGTACCAGACTCCTCCCAGAATGATTATAAT...   14.0   \n",
       "3  Abro-21      19  TATTTGTCCACCTTTTGCCGACCGAGTTTTCTTGCCCGAGACGCGT...   23.0   \n",
       "4  Abro-21      24  AATGTAGCCCCGAGCGCGAAGCCGTGAAGCTTCTGCGCGCCACCTC...   36.0   \n",
       "\n",
       "              species combUID ag1k_cluster multisp_cluster  outlier_genotype  \\\n",
       "0  Anopheles_brohieri    13-0     cl-13-18        cl-13-11             False   \n",
       "1  Anopheles_brohieri    17-0      cl-17-8         cl-17-5             False   \n",
       "2  Anopheles_brohieri    18-0     cl-18-21        cl-18-17             False   \n",
       "3  Anopheles_brohieri    19-0     cl-19-41        cl-19-22             False   \n",
       "4  Anopheles_brohieri    24-0      cl-24-1         cl-24-1             False   \n",
       "\n",
       "   split_alleles  \n",
       "0          False  \n",
       "1          False  \n",
       "2          False  \n",
       "3          False  \n",
       "4          False  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_data.to_csv(CLUSTERING, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
