{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot marker synteny plots using `jcvi.graphics.synteny`. \n",
    "\n",
    "Change introduced at line 173 of the submodule: `color = g.score#forward if strand == \"+\" else backward`. \n",
    "This was done to infer block color directly from column 5 of combined BED.\n",
    "\n",
    "Files generated for each of gambiae chromosomes:\n",
    "- display config. \n",
    "- blocks. Each row is a list of syntenic markers between markers in different species. `.`\n",
    "- merged bed of amplicon positions in three genomes (note that jcvi does not check for interchromosomal rearrangements)\n",
    "- extra features - start and end of gambiae chromosomes\n",
    "\n",
    "Usage of `jcvi.graphics.karyotype` for plotting all chromosomes at once also considered. It requires more convoluted blocks format, pairwise anchors, separate beds and other version of display config. It  does not represent amplicons as blocks, so any amplicons without homology will be invisible. Also, no interchromosomal rearrangements seem to exist in our dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# in \n",
    "IN_META = '../../data/panel_extended_info.csv'\n",
    "IN_BED_DIR = '../../../data/phylo_ampl/vobs_genomes/locationsByGenome/'\n",
    "IN_BED = {\n",
    "    'Agam': os.path.join(IN_BED_DIR, 'anopheles-gambiae-pestchromosomesagamp4.bed'),\n",
    "    'Afun': os.path.join(IN_BED_DIR, 'anopheles-funestus-fumozchromosomesafunf3.bed'),\n",
    "    'Aalb': os.path.join(IN_BED_DIR, 'anopheles-albimanus-steclachromosomesaalbs2.bed')\n",
    "}\n",
    "# intermediates\n",
    "# OUT_BED = 'data/{}.bed'\n",
    "# ANCHORS = 'data/Agam.{}.anchors.txt'\n",
    "# SEQIDS = 'data/seqids.txt'\n",
    "DISP_CONF = 'data/disp.{}.txt'\n",
    "BLOCKS = 'data/blocks.{}.txt'\n",
    "COMB_BED = 'data/comb.{}.bed'\n",
    "CHROM_ENDS = 'data/chrom_ends.{}.bed'\n",
    "# params\n",
    "SPP = ['Agam','Aalb','Afun']\n",
    "SP_CAPTIONS = {\n",
    "    'Agam': 'gambiae',\n",
    "    'Afun': 'funestus',\n",
    "    'Aalb': 'albimanus'\n",
    "}\n",
    "ANNOT_COLORS = {\n",
    "    'exon':'orangered',\n",
    "    'intron':'gold',\n",
    "    'intergenic':'royalblue'\n",
    "}\n",
    "# agam, afun, aalb\n",
    "CHROM_ENDS = {\n",
    "    \"2L\":((\"Agam_2L\",1,49364325),(\"Afun_3\",1,45633784),(\"Aalb_3L\",1,33601328)), \n",
    "    \"2R\":((\"Agam_2R\",1,61545105),(\"Afun_2\",1,56047416),(\"Aalb_2R\",1,51801758)), \n",
    "    \"3L\":((\"Agam_3L\",1,41963435),(\"Afun_3\",45633828,93833897),(\"Aalb_3R\",1,34682003)), \n",
    "    \"3R\":((\"Agam_3R\",1,53200684),(\"Afun_2\",56047552,99479338),(\"Aalb_2L\",1,37976048)), \n",
    "    \"X\":((\"Agam_X\",1,24393108),(\"Afun_X\",1,17661987))\n",
    "}\n",
    "AGAM_CHROM_SIZES = {k:v[0][2] for (k,v) in CHROM_ENDS.items()}\n",
    "# funestus chromosome arms merged in assembly\n",
    "GAM_TO_FUN_CHROM = {\n",
    "    \"2L\":\"3R\", \n",
    "    \"2R\":\"2R\", \n",
    "    \"3L\":\"3L\", \n",
    "    \"3R\":\"2L\", \n",
    "    \"X\":\"X\"\n",
    "}\n",
    "# size of drawn region for each amplicon\n",
    "REGSIZE = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read metadata\n",
    "meta = pd.read_csv(IN_META)\n",
    "# need to recode amplicon IDs from SeekDeep to current\n",
    "id_recode = meta \\\n",
    "        .set_index('SeekDeep_ID')['Primary_ID'] \\\n",
    "        .to_dict()\n",
    "id_recode['V0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Primary_ID</th>\n",
       "      <th>ID</th>\n",
       "      <th>SeekDeep_ID</th>\n",
       "      <th>SNP</th>\n",
       "      <th>F</th>\n",
       "      <th>R</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>F_deg</th>\n",
       "      <th>...</th>\n",
       "      <th>BDGP6_s_span</th>\n",
       "      <th>CpipJ2_identity</th>\n",
       "      <th>CpipJ2_q_span</th>\n",
       "      <th>CpipJ2_s_span</th>\n",
       "      <th>Comments</th>\n",
       "      <th>Outgroup ampl</th>\n",
       "      <th>conc_multiplier</th>\n",
       "      <th>annotation</th>\n",
       "      <th>start_insert</th>\n",
       "      <th>end_insert</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>31_var</td>\n",
       "      <td>V0</td>\n",
       "      <td>high freq</td>\n",
       "      <td>TGTSTACGGTCTGAAGAACATc</td>\n",
       "      <td>TTATCCGGCTCCAAGTTAAGG</td>\n",
       "      <td>2L</td>\n",
       "      <td>10356417</td>\n",
       "      <td>10356608</td>\n",
       "      <td>TGT[G/C]TACGGTCTGAAGAACATc</td>\n",
       "      <td>...</td>\n",
       "      <td>2R:22800075-22800136</td>\n",
       "      <td>0.968253968</td>\n",
       "      <td>1-63</td>\n",
       "      <td>supercont3.30:1160215-1160153</td>\n",
       "      <td>F primer perfect match in Cpip&amp;Aaeg</td>\n",
       "      <td>No</td>\n",
       "      <td>10</td>\n",
       "      <td>intron</td>\n",
       "      <td>10356439</td>\n",
       "      <td>10356587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>33_other</td>\n",
       "      <td>O0</td>\n",
       "      <td>low freq</td>\n",
       "      <td>GAGCGtGCGGCcAAGATG</td>\n",
       "      <td>ACAgACCGACGTTAATGGC</td>\n",
       "      <td>2L</td>\n",
       "      <td>10357458</td>\n",
       "      <td>10357648</td>\n",
       "      <td>GAGCGtGCGGCcAAGATG</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.926315789</td>\n",
       "      <td>1-190</td>\n",
       "      <td>supercont3.30:1159280-1159091</td>\n",
       "      <td>Cpip - perfect, Aaeg - 5` subst</td>\n",
       "      <td>Yes</td>\n",
       "      <td>10</td>\n",
       "      <td>exon</td>\n",
       "      <td>10357476</td>\n",
       "      <td>10357629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>44_other</td>\n",
       "      <td>O1</td>\n",
       "      <td>low freq</td>\n",
       "      <td>CAGTCAAATTTCCAGACAATCT</td>\n",
       "      <td>CGGAAGTGCATTTGAAGGAAaA</td>\n",
       "      <td>2L</td>\n",
       "      <td>12084495</td>\n",
       "      <td>12084690</td>\n",
       "      <td>CAGTCAAATTTCCAGACAATCT</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.8</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>12084517</td>\n",
       "      <td>12084668</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>47_other</td>\n",
       "      <td>O2</td>\n",
       "      <td>low freq</td>\n",
       "      <td>GaTATAAATTGTCGATCACACAAACT</td>\n",
       "      <td>TGCATTTATCGTAGTACAATCTCA</td>\n",
       "      <td>2L</td>\n",
       "      <td>13862864</td>\n",
       "      <td>13863055</td>\n",
       "      <td>GaTATAAATTGTCGATCACACAAACT</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.6</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>13862890</td>\n",
       "      <td>13863031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>58_cons</td>\n",
       "      <td>C0</td>\n",
       "      <td>high freq</td>\n",
       "      <td>ATGcTBGTCATgATGATGATCT</td>\n",
       "      <td>CCGATCCACGATAAGGAGTAC</td>\n",
       "      <td>2L</td>\n",
       "      <td>18693481</td>\n",
       "      <td>18693714</td>\n",
       "      <td>ATGcT[C/T/G]GTCATgATGATGATCT</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.834042553</td>\n",
       "      <td>1-233</td>\n",
       "      <td>supercont3.137:804142-804374</td>\n",
       "      <td>5`subst in Aaeg&amp;Cpip</td>\n",
       "      <td>Yes</td>\n",
       "      <td>10</td>\n",
       "      <td>exon</td>\n",
       "      <td>18693503</td>\n",
       "      <td>18693693</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Primary_ID        ID SeekDeep_ID        SNP                           F  \\\n",
       "0          0    31_var          V0  high freq      TGTSTACGGTCTGAAGAACATc   \n",
       "1          1  33_other          O0   low freq          GAGCGtGCGGCcAAGATG   \n",
       "2          2  44_other          O1   low freq      CAGTCAAATTTCCAGACAATCT   \n",
       "3          3  47_other          O2   low freq  GaTATAAATTGTCGATCACACAAACT   \n",
       "4          4   58_cons          C0  high freq      ATGcTBGTCATgATGATGATCT   \n",
       "\n",
       "                          R chr     start       end  \\\n",
       "0     TTATCCGGCTCCAAGTTAAGG  2L  10356417  10356608   \n",
       "1       ACAgACCGACGTTAATGGC  2L  10357458  10357648   \n",
       "2    CGGAAGTGCATTTGAAGGAAaA  2L  12084495  12084690   \n",
       "3  TGCATTTATCGTAGTACAATCTCA  2L  13862864  13863055   \n",
       "4     CCGATCCACGATAAGGAGTAC  2L  18693481  18693714   \n",
       "\n",
       "                          F_deg  ...          BDGP6_s_span CpipJ2_identity  \\\n",
       "0    TGT[G/C]TACGGTCTGAAGAACATc  ...  2R:22800075-22800136     0.968253968   \n",
       "1            GAGCGtGCGGCcAAGATG  ...                   NaN     0.926315789   \n",
       "2        CAGTCAAATTTCCAGACAATCT  ...                   NaN             NaN   \n",
       "3    GaTATAAATTGTCGATCACACAAACT  ...                   NaN             NaN   \n",
       "4  ATGcT[C/T/G]GTCATgATGATGATCT  ...                   NaN     0.834042553   \n",
       "\n",
       "   CpipJ2_q_span                  CpipJ2_s_span  \\\n",
       "0           1-63  supercont3.30:1160215-1160153   \n",
       "1          1-190  supercont3.30:1159280-1159091   \n",
       "2            NaN                            NaN   \n",
       "3            NaN                            NaN   \n",
       "4          1-233   supercont3.137:804142-804374   \n",
       "\n",
       "                              Comments  Outgroup ampl  conc_multiplier  \\\n",
       "0  F primer perfect match in Cpip&Aaeg             No               10   \n",
       "1      Cpip - perfect, Aaeg - 5` subst            Yes               10   \n",
       "2                                  NaN            NaN              5.8   \n",
       "3                                  NaN            NaN              2.6   \n",
       "4                 5`subst in Aaeg&Cpip            Yes               10   \n",
       "\n",
       "   annotation  start_insert  end_insert  \n",
       "0      intron      10356439    10356587  \n",
       "1        exon      10357476    10357629  \n",
       "2  intergenic      12084517    12084668  \n",
       "3  intergenic      13862890    13863031  \n",
       "4        exon      18693503    18693693  \n",
       "\n",
       "[5 rows x 49 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta['annotation'] = (meta.intergenic == 'Yes').replace({True:'intergenic',False:'exon'})\n",
    "meta.loc[meta.intron == 'Yes', 'annotation'] = 'intron'\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'intron'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot = meta.set_index('Primary_ID')['annotation'].to_dict()\n",
    "annot['0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Primary_ID,ID,SeekDeep_ID,SNP,F,R,chr,start,end,F_deg,R_deg,identical_seq,idenified_lineages,aligned_spp,unaligned_spp,aligned_ampl_length,agam_ampl_length,aligned_insert_length,agam_insert_length,total_snvs,total_indels,insert_snvs,insert_indels,prop_id_lineages,exon,gene,intergenic,intron,mRNA,repeat,utr,AaegL5_identity,AaegL5_q_span,AaegL5_s_span,AgamP3_identity,AgamP3_q_span,AgamP3_s_span,BDGP6_identity,BDGP6_q_span,BDGP6_s_span,CpipJ2_identity,CpipJ2_q_span,CpipJ2_s_span,Comments,Outgroup ampl,conc_multiplier,annotation,start_insert,end_insert\r\n",
      "0,31_var,V0,high freq,TGTSTACGGTCTGAAGAACATc,TTATCCGGCTCCAAGTTAAGG,2L,10356417,10356608,TGT[G/C]TACGGTCTGAAGAACATc,TT[A/G]TCCGGCTCCAAGTTAAGG,\"[{'AsteI2', 'AsteS1'}]\",20.0,21.0,,259,191.0,216.0,148.0,33.0,139.0,30.0,139.0,0.9523809520000001,E016098A;E016099A,AGAP005134,No,Yes,AGAP005134-RA,None,No,0.96969697,1-66,2:204342881-204342946,1.0,1-191,2L:10356418-10356608,0.951612903,1-62,2R:22800075-22800136,0.968253968,1-63,supercont3.30:1160215-1160153,F primer perfect match in Cpip&Aaeg,No,10,intron,10356439,10356587\r\n"
     ]
    }
   ],
   "source": [
    "# write annotations to original metadata\n",
    "meta.to_csv(IN_META, index=False)\n",
    "! head -2 {IN_META}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>len</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>2538226</td>\n",
       "      <td>2538411</td>\n",
       "      <td>Afun_12</td>\n",
       "      <td>185</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>10269306</td>\n",
       "      <td>10269539</td>\n",
       "      <td>Afun_16</td>\n",
       "      <td>233</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>18159344</td>\n",
       "      <td>18159529</td>\n",
       "      <td>Afun_25</td>\n",
       "      <td>185</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>18179529</td>\n",
       "      <td>18179730</td>\n",
       "      <td>Afun_24</td>\n",
       "      <td>201</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>23084797</td>\n",
       "      <td>23084987</td>\n",
       "      <td>Afun_22</td>\n",
       "      <td>190</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom     start       end       id  len strand\n",
       "0     2   2538226   2538411  Afun_12  185      +\n",
       "1     2  10269306  10269539  Afun_16  233      +\n",
       "2     2  18159344  18159529  Afun_25  185      -\n",
       "3     2  18179529  18179730  Afun_24  201      -\n",
       "4     2  23084797  23084987  Afun_22  190      -"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# process input BEDs\n",
    "comb_data = dict()\n",
    "for sp, bedfile in IN_BED.items():\n",
    "    sp_data = pd.read_csv(bedfile, \n",
    "                          sep='\\t', \n",
    "                          header=None,\n",
    "                          names=['chrom','start','end','id','len','strand'])\n",
    "    # get amplicon name\n",
    "    sd_ids = sp_data['id'].str.split('=').str.get(-1).str.slice(0,-1)\n",
    "    # get extraction count\n",
    "    extr_count = sp_data['id'].str.split('=').str.get(1).str.split(';').str.get(0)\n",
    "    # replace and prepend with species name\n",
    "    sp_data['id'] = sp + '_' + sd_ids.replace(id_recode)\n",
    "    # remove repetitive positions (if any)\n",
    "    sp_data['extraction'] = extr_count\n",
    "    sp_data = sp_data[sp_data['extraction'] == '0'].drop(columns=['extraction'])\n",
    "    comb_data[sp] = sp_data\n",
    "comb_data['Afun'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>len</th>\n",
       "      <th>strand</th>\n",
       "      <th>species</th>\n",
       "      <th>amplicon</th>\n",
       "      <th>annotation</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aalb_2L</td>\n",
       "      <td>1067926</td>\n",
       "      <td>1068146</td>\n",
       "      <td>Aalb_53</td>\n",
       "      <td>220</td>\n",
       "      <td>-</td>\n",
       "      <td>Aalb</td>\n",
       "      <td>53</td>\n",
       "      <td>exon</td>\n",
       "      <td>orangered</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aalb_2L</td>\n",
       "      <td>5419554</td>\n",
       "      <td>5419760</td>\n",
       "      <td>Aalb_47</td>\n",
       "      <td>206</td>\n",
       "      <td>+</td>\n",
       "      <td>Aalb</td>\n",
       "      <td>47</td>\n",
       "      <td>intron</td>\n",
       "      <td>gold</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aalb_2L</td>\n",
       "      <td>8671327</td>\n",
       "      <td>8671500</td>\n",
       "      <td>Aalb_44</td>\n",
       "      <td>173</td>\n",
       "      <td>-</td>\n",
       "      <td>Aalb</td>\n",
       "      <td>44</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>royalblue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aalb_2L</td>\n",
       "      <td>10922763</td>\n",
       "      <td>10922958</td>\n",
       "      <td>Aalb_40</td>\n",
       "      <td>195</td>\n",
       "      <td>-</td>\n",
       "      <td>Aalb</td>\n",
       "      <td>40</td>\n",
       "      <td>intron</td>\n",
       "      <td>gold</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aalb_2L</td>\n",
       "      <td>21336966</td>\n",
       "      <td>21337166</td>\n",
       "      <td>Aalb_39</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "      <td>Aalb</td>\n",
       "      <td>39</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>royalblue</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end       id  len strand species  amplicon  \\\n",
       "0  Aalb_2L   1067926   1068146  Aalb_53  220      -    Aalb        53   \n",
       "1  Aalb_2L   5419554   5419760  Aalb_47  206      +    Aalb        47   \n",
       "2  Aalb_2L   8671327   8671500  Aalb_44  173      -    Aalb        44   \n",
       "3  Aalb_2L  10922763  10922958  Aalb_40  195      -    Aalb        40   \n",
       "4  Aalb_2L  21336966  21337166  Aalb_39  200      -    Aalb        39   \n",
       "\n",
       "   annotation      color  \n",
       "0        exon  orangered  \n",
       "1      intron       gold  \n",
       "2  intergenic  royalblue  \n",
       "3      intron       gold  \n",
       "4  intergenic  royalblue  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# additional data\n",
    "comb_data = pd.concat(comb_data.values())\n",
    "comb_data.shape\n",
    "sa = comb_data['id'].str.split('_', expand=True)\n",
    "comb_data['species'] = sa[0]\n",
    "comb_data['amplicon'] = sa[1].astype(int)\n",
    "comb_data['annotation'] = sa[1].replace(annot)\n",
    "comb_data['color'] = comb_data['annotation'].replace(ANNOT_COLORS)\n",
    "# chromosome names to be unique between species\n",
    "comb_data['chrom'] = comb_data['species'] + '_' + comb_data['chrom']\n",
    "# shift\n",
    "comb_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare input files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# test chromosome\n",
    "chrom = '2L'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_tsv(f, d):\n",
    "    '''common writing - beds and blocks'''\n",
    "    d.to_csv(f,\n",
    "            sep='\\t',\n",
    "            header=False,\n",
    "            index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      0\n",
       "1      1\n",
       "2      2\n",
       "3      3\n",
       "4      4\n",
       "5      5\n",
       "6      6\n",
       "7      7\n",
       "8      8\n",
       "9      9\n",
       "10    10\n",
       "11    11\n",
       "Name: amplicon, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get amplicons on chromosome\n",
    "chrom_ampls = comb_data.loc[comb_data.chrom == 'Agam_' + chrom, 'amplicon']\n",
    "chrom_ampls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(27, 10)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset data to chromosome\n",
    "chrom_data = comb_data[comb_data.amplicon.isin(chrom_ampls)].copy()\n",
    "chrom_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>corr_start</th>\n",
       "      <th>corr_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10356417</td>\n",
       "      <td>10356608</td>\n",
       "      <td>9357032</td>\n",
       "      <td>10257032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10357458</td>\n",
       "      <td>10357648</td>\n",
       "      <td>10357032</td>\n",
       "      <td>11257032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12084495</td>\n",
       "      <td>12084690</td>\n",
       "      <td>11584495</td>\n",
       "      <td>12584690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13862864</td>\n",
       "      <td>13863055</td>\n",
       "      <td>13362864</td>\n",
       "      <td>14363055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18693481</td>\n",
       "      <td>18693714</td>\n",
       "      <td>17879422</td>\n",
       "      <td>18779422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>19065171</td>\n",
       "      <td>19065363</td>\n",
       "      <td>18879422</td>\n",
       "      <td>19779422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>20840388</td>\n",
       "      <td>20840620</td>\n",
       "      <td>20340388</td>\n",
       "      <td>21340620</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>26760700</td>\n",
       "      <td>26760890</td>\n",
       "      <td>26217768</td>\n",
       "      <td>27117768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>27674646</td>\n",
       "      <td>27674836</td>\n",
       "      <td>27217768</td>\n",
       "      <td>28117768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>40854755</td>\n",
       "      <td>40854948</td>\n",
       "      <td>40354755</td>\n",
       "      <td>41354948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>45852869</td>\n",
       "      <td>45853061</td>\n",
       "      <td>45352869</td>\n",
       "      <td>46353061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>48612516</td>\n",
       "      <td>48612730</td>\n",
       "      <td>48112516</td>\n",
       "      <td>49112730</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       start       end  corr_start  corr_end\n",
       "0   10356417  10356608     9357032  10257032\n",
       "1   10357458  10357648    10357032  11257032\n",
       "2   12084495  12084690    11584495  12584690\n",
       "3   13862864  13863055    13362864  14363055\n",
       "4   18693481  18693714    17879422  18779422\n",
       "5   19065171  19065363    18879422  19779422\n",
       "6   20840388  20840620    20340388  21340620\n",
       "7   26760700  26760890    26217768  27117768\n",
       "8   27674646  27674836    27217768  28117768\n",
       "9   40854755  40854948    40354755  41354948\n",
       "10  45852869  45853061    45352869  46353061\n",
       "11  48612516  48612730    48112516  49112730"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# make blocks longer for visualisation, resolve overlaps\n",
    "def longer_blocks(starts, ends, target_len, chrom_len=None):\n",
    "    '''\n",
    "    Make blocks longer shifting left if overlapping\n",
    "    '''\n",
    "    \n",
    "    shift = int(target_len / 2)\n",
    "    # gap introduced for when plotting overlapping regions\n",
    "    ovl_gap = 0.1\n",
    "        \n",
    "    # initial enlargement\n",
    "    corr_starts = starts - shift\n",
    "    corr_ends = ends + shift\n",
    "    \n",
    "    # overlap removal\n",
    "    nov_starts = list()\n",
    "    nov_ends = list()\n",
    "    ov = list()\n",
    "    for s,e in zip(corr_starts, corr_ends):\n",
    "        # init overlap analysis\n",
    "        if len(ov) == 0:\n",
    "            ov = [[s,e]]\n",
    "            continue\n",
    "        # no overlap: save previous region, store current\n",
    "        if s >= ov[-1][-1]:\n",
    "            for (novs, nove) in ov:\n",
    "                nov_starts.append(novs)\n",
    "                nov_ends.append(nove)\n",
    "            ov = [[s,e]]\n",
    "        # overlap - add current region, spread\n",
    "        else:\n",
    "            ov.append([s,e])\n",
    "            # coordinates of the resulting stack of overlapping regions\n",
    "            center = (ov[0][0] + ov[-1][-1]) / 2\n",
    "            # adjust to chromosome ends\n",
    "            center = max(center, shift * len(ov) + 1)\n",
    "            if chrom_len:\n",
    "                center = min(center, chrom_len - shift * len(ov))\n",
    "            new_start = int(center) - shift * len(ov)\n",
    "            # apply shift, leaving small gap between amplicons\n",
    "            for i in range(len(ov)):\n",
    "                # start\n",
    "                ov[i][0] = int(new_start + shift * i * 2)\n",
    "                # end \n",
    "                \n",
    "                ov[i][1] = int(new_start + shift * (i + 1 - ovl_gap) * 2)\n",
    "    # store remaining coordinates\n",
    "    for (novs, nove) in ov:\n",
    "        nov_starts.append(novs)\n",
    "        nov_ends.append(nove)\n",
    "    \n",
    "    # trim starts and ends to fit chromosome\n",
    "#     if nov_starts[0] < 1:\n",
    "#         nov_starts[0] = 1\n",
    "#     if chrom_len:\n",
    "#         if nov_ends[-1] > chrom_len:\n",
    "#             nov_ends[-1] = chrom_len\n",
    "    \n",
    "    return (nov_starts, nov_ends)\n",
    "gam_cdata = chrom_data[chrom_data.species=='Agam'].copy()\n",
    "\n",
    "corr_coords = longer_blocks(gam_cdata['start'],\n",
    "                            gam_cdata['end'],\n",
    "                            REGSIZE,\n",
    "                            AGAM_CHROM_SIZES[chrom])\n",
    "gam_cdata['corr_start'] = corr_coords[0]\n",
    "gam_cdata['corr_end'] = corr_coords[1]\n",
    "display(gam_cdata[['start','end','corr_start','corr_end']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>len</th>\n",
       "      <th>strand</th>\n",
       "      <th>species</th>\n",
       "      <th>amplicon</th>\n",
       "      <th>annotation</th>\n",
       "      <th>color</th>\n",
       "      <th>corr_start</th>\n",
       "      <th>corr_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Agam_2L</td>\n",
       "      <td>10356417</td>\n",
       "      <td>10356608</td>\n",
       "      <td>Agam_0</td>\n",
       "      <td>191</td>\n",
       "      <td>+</td>\n",
       "      <td>Agam</td>\n",
       "      <td>0</td>\n",
       "      <td>intron</td>\n",
       "      <td>gold</td>\n",
       "      <td>9357032</td>\n",
       "      <td>10257032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Agam_2L</td>\n",
       "      <td>10357458</td>\n",
       "      <td>10357648</td>\n",
       "      <td>Agam_1</td>\n",
       "      <td>190</td>\n",
       "      <td>+</td>\n",
       "      <td>Agam</td>\n",
       "      <td>1</td>\n",
       "      <td>exon</td>\n",
       "      <td>orangered</td>\n",
       "      <td>10357032</td>\n",
       "      <td>11257032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Agam_2L</td>\n",
       "      <td>12084495</td>\n",
       "      <td>12084690</td>\n",
       "      <td>Agam_2</td>\n",
       "      <td>195</td>\n",
       "      <td>+</td>\n",
       "      <td>Agam</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>royalblue</td>\n",
       "      <td>11584495</td>\n",
       "      <td>12584690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Agam_2L</td>\n",
       "      <td>13862864</td>\n",
       "      <td>13863055</td>\n",
       "      <td>Agam_3</td>\n",
       "      <td>191</td>\n",
       "      <td>+</td>\n",
       "      <td>Agam</td>\n",
       "      <td>3</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>royalblue</td>\n",
       "      <td>13362864</td>\n",
       "      <td>14363055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Agam_2L</td>\n",
       "      <td>18693481</td>\n",
       "      <td>18693714</td>\n",
       "      <td>Agam_4</td>\n",
       "      <td>233</td>\n",
       "      <td>+</td>\n",
       "      <td>Agam</td>\n",
       "      <td>4</td>\n",
       "      <td>exon</td>\n",
       "      <td>orangered</td>\n",
       "      <td>17879422</td>\n",
       "      <td>18779422</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end      id  len strand species  amplicon  \\\n",
       "0  Agam_2L  10356417  10356608  Agam_0  191      +    Agam         0   \n",
       "1  Agam_2L  10357458  10357648  Agam_1  190      +    Agam         1   \n",
       "2  Agam_2L  12084495  12084690  Agam_2  195      +    Agam         2   \n",
       "3  Agam_2L  13862864  13863055  Agam_3  191      +    Agam         3   \n",
       "4  Agam_2L  18693481  18693714  Agam_4  233      +    Agam         4   \n",
       "\n",
       "   annotation      color  corr_start  corr_end  \n",
       "0      intron       gold     9357032  10257032  \n",
       "1        exon  orangered    10357032  11257032  \n",
       "2  intergenic  royalblue    11584495  12584690  \n",
       "3  intergenic  royalblue    13362864  14363055  \n",
       "4        exon  orangered    17879422  18779422  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# apply shift to all species\n",
    "def longer_allsp(cdata, chrom_len):\n",
    "    corr_data = list()\n",
    "    for sp in SPP:\n",
    "        clen = chrom_len if sp == 'Agam' else None\n",
    "        sp_chrom_data = chrom_data[chrom_data.species==sp].sort_values('start')\n",
    "        # do not parse empty coordinate lists\n",
    "        if sp_chrom_data.shape[0] > 0:\n",
    "            corr_coords = longer_blocks(sp_chrom_data['start'],\n",
    "                                        sp_chrom_data['end'],\n",
    "                                        REGSIZE,\n",
    "                                        clen)\n",
    "            sp_chrom_data['corr_start'] = corr_coords[0]\n",
    "            sp_chrom_data['corr_end'] = corr_coords[1]\n",
    "            corr_data.append(sp_chrom_data)\n",
    "    return pd.concat(corr_data)\n",
    "corr_data = longer_allsp(chrom_data, AGAM_CHROM_SIZES[chrom])\n",
    "corr_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>id</th>\n",
       "      <th>len</th>\n",
       "      <th>strand</th>\n",
       "      <th>species</th>\n",
       "      <th>amplicon</th>\n",
       "      <th>annotation</th>\n",
       "      <th>color</th>\n",
       "      <th>corr_start</th>\n",
       "      <th>corr_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Afun_3</td>\n",
       "      <td>18114432.0</td>\n",
       "      <td>18114622.0</td>\n",
       "      <td>Afun_6</td>\n",
       "      <td>190.0</td>\n",
       "      <td>-</td>\n",
       "      <td>Afun</td>\n",
       "      <td>6</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>royalblue</td>\n",
       "      <td>17614432</td>\n",
       "      <td>18614622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Afun_3</td>\n",
       "      <td>25816570.0</td>\n",
       "      <td>25816748.0</td>\n",
       "      <td>Afun_8</td>\n",
       "      <td>178.0</td>\n",
       "      <td>+</td>\n",
       "      <td>Afun</td>\n",
       "      <td>8</td>\n",
       "      <td>intron</td>\n",
       "      <td>gold</td>\n",
       "      <td>25316570</td>\n",
       "      <td>26316748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Afun_3</td>\n",
       "      <td>28805613.0</td>\n",
       "      <td>28805828.0</td>\n",
       "      <td>Afun_7</td>\n",
       "      <td>215.0</td>\n",
       "      <td>+</td>\n",
       "      <td>Afun</td>\n",
       "      <td>7</td>\n",
       "      <td>intron</td>\n",
       "      <td>gold</td>\n",
       "      <td>28305613</td>\n",
       "      <td>29305828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Afun_3</td>\n",
       "      <td>41162360.0</td>\n",
       "      <td>41162521.0</td>\n",
       "      <td>Afun_2</td>\n",
       "      <td>161.0</td>\n",
       "      <td>-</td>\n",
       "      <td>Afun</td>\n",
       "      <td>2</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>royalblue</td>\n",
       "      <td>40662360</td>\n",
       "      <td>41662521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Agam_2L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Agam_start</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Agam</td>\n",
       "      <td>998</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gainsboro</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Agam_2L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Agam_end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Agam</td>\n",
       "      <td>999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gainsboro</td>\n",
       "      <td>49364324</td>\n",
       "      <td>49364325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Afun_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Afun_start</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Afun</td>\n",
       "      <td>996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gainsboro</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Afun_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Afun_end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Afun</td>\n",
       "      <td>997</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gainsboro</td>\n",
       "      <td>45633783</td>\n",
       "      <td>45633784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Aalb_3L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aalb_start</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aalb</td>\n",
       "      <td>994</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gainsboro</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Aalb_3L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aalb_end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Aalb</td>\n",
       "      <td>995</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gainsboro</td>\n",
       "      <td>33601327</td>\n",
       "      <td>33601328</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom       start         end          id    len strand species  \\\n",
       "33   Afun_3  18114432.0  18114622.0      Afun_6  190.0      -    Afun   \n",
       "35   Afun_3  25816570.0  25816748.0      Afun_8  178.0      +    Afun   \n",
       "36   Afun_3  28805613.0  28805828.0      Afun_7  215.0      +    Afun   \n",
       "37   Afun_3  41162360.0  41162521.0      Afun_2  161.0      -    Afun   \n",
       "0   Agam_2L         NaN         NaN  Agam_start    NaN    NaN    Agam   \n",
       "1   Agam_2L         NaN         NaN    Agam_end    NaN    NaN    Agam   \n",
       "2    Afun_3         NaN         NaN  Afun_start    NaN    NaN    Afun   \n",
       "3    Afun_3         NaN         NaN    Afun_end    NaN    NaN    Afun   \n",
       "4   Aalb_3L         NaN         NaN  Aalb_start    NaN    NaN    Aalb   \n",
       "5   Aalb_3L         NaN         NaN    Aalb_end    NaN    NaN    Aalb   \n",
       "\n",
       "    amplicon  annotation      color  corr_start  corr_end  \n",
       "33         6  intergenic  royalblue    17614432  18614622  \n",
       "35         8      intron       gold    25316570  26316748  \n",
       "36         7      intron       gold    28305613  29305828  \n",
       "37         2  intergenic  royalblue    40662360  41662521  \n",
       "0        998         NaN  gainsboro           1         2  \n",
       "1        999         NaN  gainsboro    49364324  49364325  \n",
       "2        996         NaN  gainsboro           1         2  \n",
       "3        997         NaN  gainsboro    45633783  45633784  \n",
       "4        994         NaN  gainsboro           1         2  \n",
       "5        995         NaN  gainsboro    33601327  33601328  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add chromosome ends\n",
    "def add_ce(cdata, chrom, chrom_data):\n",
    "    color = 'gainsboro'\n",
    "    ce = []\n",
    "    for i, (chrom, start, end) in enumerate(chrom_data):\n",
    "        species = chrom.split('_')[0]\n",
    "        ce.extend([\n",
    "            # start \n",
    "            {'chrom':chrom,\n",
    "            'corr_start':start,\n",
    "            'corr_end':start + 1,\n",
    "            'id':species + '_start',\n",
    "            'amplicon':1000 - 2 * (i + 1),\n",
    "            'species':species,\n",
    "            'color':color},\n",
    "            # end\n",
    "            {'chrom':chrom,\n",
    "            'corr_start':end - 1,\n",
    "            'corr_end':end,\n",
    "            'id':species + '_end',\n",
    "            'amplicon':1000 - 2 * (i + 1) + 1,\n",
    "            'species':species,\n",
    "            'color':color}\n",
    "        ])\n",
    "    return cdata.append(pd.DataFrame(ce), sort=False)\n",
    "corr_data = add_ce(corr_data, chrom, CHROM_ENDS[chrom])\n",
    "corr_data.tail(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BED files\n",
    "to_tsv(COMB_BED.format(chrom),\n",
    "       corr_data[['chrom','corr_start','corr_end','id','color']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gainsboro*Agam_0\tAalb_0\t.\r\n",
      "gainsboro*Agam_1\t.\t.\r\n",
      "gainsboro*Agam_2\t.\tAfun_2\r\n",
      "gainsboro*Agam_3\tAalb_3\tAfun_3\r\n",
      "gainsboro*Agam_4\tAalb_4\tAfun_4\r\n",
      "gainsboro*Agam_5\t.\tAfun_5\r\n",
      "gainsboro*Agam_6\t.\tAfun_6\r\n",
      "gainsboro*Agam_7\tAalb_7\tAfun_7\r\n",
      "gainsboro*Agam_8\tAalb_8\tAfun_8\r\n",
      "gainsboro*Agam_9\tAalb_9\t.\r\n"
     ]
    }
   ],
   "source": [
    "# multiway blocks\n",
    "def write_blocks(cdata, chrom):\n",
    "    block_data = cdata.pivot(index='amplicon', columns='species', values='id') \\\n",
    "                    .fillna('.')\n",
    "    aligned_spp = [sp for sp in SPP if sp in block_data.columns]\n",
    "    block_data = block_data[aligned_spp]\n",
    "    # highlight thy synteny relations, otherwise the markers will be dim\n",
    "    block_data['Agam'] = 'gainsboro*' + block_data['Agam']\n",
    "    to_tsv(BLOCKS.format(chrom),\n",
    "           block_data)\n",
    "write_blocks(corr_data, chrom)\n",
    "! head {BLOCKS.format(chrom)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# x,   y, rotation,     ha,     va, color, ratio,            label\r\n",
      "0.5, 0.5,        0, center,    top,      k,     0.802083691303,       An. gambiae 2L\r\n",
      "0.5, 0.65,        0, center, top,      k,     0.802083691303,       An. albimanus 3L\r\n",
      "0.5, 0.35,        0, center, bottom,      k,     0.802083691303,        An. funestus 3R\r\n",
      "# edges\r\n",
      "e, 0, 1\r\n",
      "e, 0, 2\r\n"
     ]
    }
   ],
   "source": [
    "# display config\n",
    "def write_disp(cdata, chrom):\n",
    "    \n",
    "    # scale all chromosomes relative to max\n",
    "    ratio = float(AGAM_CHROM_SIZES[chrom]) / max(AGAM_CHROM_SIZES.values())\n",
    "    # get chromosomes for other species, assume all markers on the same chromsome\n",
    "    # funestus assembly does not have arms, use pre-defined arm mapping\n",
    "    fun_chrom = GAM_TO_FUN_CHROM[chrom]\n",
    "    # all species present in alignment \n",
    "    try:\n",
    "        # albimanus chromosome from assembly header\n",
    "        alb_chrom = cdata.loc[cdata.species == 'Aalb','chrom'].iloc[0].split('_')[-1]\n",
    "        with open(DISP_CONF.format(chrom), 'w') as o:\n",
    "            o.write(\"\"\"# x,   y, rotation,     ha,     va, color, ratio,            label\n",
    "0.5, 0.5,        0, center,    top,      k,     {ratio},       An. gambiae {}\n",
    "0.5, 0.65,        0, center, top,      k,     {ratio},       An. albimanus {}\n",
    "0.5, 0.35,        0, center, bottom,      k,     {ratio},        An. funestus {}\n",
    "# edges\n",
    "e, 0, 1\n",
    "e, 0, 2\n",
    "\"\"\".format(chrom, alb_chrom, fun_chrom, ratio=ratio))\n",
    "    # homology loss happens for alb X\n",
    "    except:\n",
    "        with open(DISP_CONF.format(chrom), 'w') as o:\n",
    "            o.write(\"\"\"# x,   y, rotation,     ha,     va, color, ratio,            label\n",
    "0.5, 0.5,        0, center,    top,      k,     {ratio},       An. gambiae {}\n",
    "0.5, 0.35,        0, center, bottom,      k,     {ratio},        An. funestus {}\n",
    "# edges\n",
    "e, 0, 1\n",
    "\"\"\".format(chrom, fun_chrom, ratio=ratio))\n",
    "write_disp(corr_data, chrom)\n",
    "! head {DISP_CONF.format(chrom)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run visualisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0;33m13:06:37 [base]\u001b[0m\u001b[0;35m Load file `data/comb.2L.bed`\u001b[0m\n",
      "\u001b[0;33m13:06:37 [base]\u001b[0m\u001b[0;35m Load file `data/blocks.2L.txt`\u001b[0m\n",
      "\u001b[0;33m13:06:37 [base]\u001b[0m\u001b[0;35m Load file `data/disp.2L.txt`\u001b[0m\n",
      "Column 0: Agam_start - Agam_end (Agam_2L:2-49364325)\n",
      "  Agam_2L .. 14 (14) features .. +\n",
      "Column 1: Aalb_start - Aalb_end (Aalb_3L:2-33601328)\n",
      "  Aalb_3L .. 9 (9) features .. -\n",
      "Column 2: Afun_start - Afun_end (Afun_3:2-45633784)\n",
      "  Afun_3 .. 10 (10) features .. -\n",
      "/Users/am60/malaria/malariagen/binder/conda/envs/75_synteny_plot/lib/python2.7/site-packages/matplotlib/font_manager.py:1333: UserWarning: findfont: Font family [u'sans-serif'] not found. Falling back to DejaVu Sans\n",
      "  (prop.get_family(), self.defaultFamily[fontext]))\n",
      "\u001b[0;33m13:06:38 [base]\u001b[0m\u001b[0;35m Figure saved to `data/blocks.2L.pdf` (2400px x 2100px)\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "! python -m jcvi.graphics.synteny --notex \\\n",
    "    {BLOCKS.format(chrom)} \\\n",
    "    {COMB_BED.format(chrom)} \\\n",
    "    {DISP_CONF.format(chrom)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Apply to all chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0;33m13:06:39 [base]\u001b[0m\u001b[0;35m Load file `data/comb.2L.bed`\u001b[0m\n",
      "\u001b[0;33m13:06:39 [base]\u001b[0m\u001b[0;35m Load file `data/blocks.2L.txt`\u001b[0m\n",
      "\u001b[0;33m13:06:39 [base]\u001b[0m\u001b[0;35m Load file `data/disp.2L.txt`\u001b[0m\n",
      "Column 0: Agam_start - Agam_end (Agam_2L:2-49364325)\n",
      "  Agam_2L .. 14 (14) features .. +\n",
      "Column 1: Aalb_start - Aalb_end (Aalb_3L:2-33601328)\n",
      "  Aalb_3L .. 9 (9) features .. -\n",
      "Column 2: Afun_start - Afun_end (Afun_3:2-45633784)\n",
      "  Afun_3 .. 10 (10) features .. -\n",
      "/Users/am60/malaria/malariagen/binder/conda/envs/75_synteny_plot/lib/python2.7/site-packages/matplotlib/font_manager.py:1333: UserWarning: findfont: Font family [u'sans-serif'] not found. Falling back to DejaVu Sans\n",
      "  (prop.get_family(), self.defaultFamily[fontext]))\n",
      "\u001b[0;33m13:06:40 [base]\u001b[0m\u001b[0;35m Figure saved to `data/blocks.2L.pdf` (2400px x 2100px)\u001b[0m\n",
      "\u001b[0;33m13:06:42 [base]\u001b[0m\u001b[0;35m Load file `data/comb.3R.bed`\u001b[0m\n",
      "\u001b[0;33m13:06:42 [base]\u001b[0m\u001b[0;35m Load file `data/blocks.3R.txt`\u001b[0m\n",
      "\u001b[0;33m13:06:42 [base]\u001b[0m\u001b[0;35m Load file `data/disp.3R.txt`\u001b[0m\n",
      "Column 0: Agam_start - Agam_end (Agam_3R:2-53200684)\n",
      "  Agam_3R .. 19 (19) features .. +\n",
      "Column 1: Aalb_start - Aalb_end (Aalb_2L:2-37976048)\n",
      "  Aalb_2L .. 11 (11) features .. -\n",
      "Column 2: Afun_start - Afun_end (Afun_2:56047553-99479338)\n",
      "  Afun_2 .. 17 (17) features .. -\n",
      "/Users/am60/malaria/malariagen/binder/conda/envs/75_synteny_plot/lib/python2.7/site-packages/matplotlib/font_manager.py:1333: UserWarning: findfont: Font family [u'sans-serif'] not found. Falling back to DejaVu Sans\n",
      "  (prop.get_family(), self.defaultFamily[fontext]))\n",
      "\u001b[0;33m13:06:42 [base]\u001b[0m\u001b[0;35m Figure saved to `data/blocks.3R.pdf` (2400px x 2100px)\u001b[0m\n",
      "\u001b[0;33m13:06:44 [base]\u001b[0m\u001b[0;35m Load file `data/comb.3L.bed`\u001b[0m\n",
      "\u001b[0;33m13:06:44 [base]\u001b[0m\u001b[0;35m Load file `data/blocks.3L.txt`\u001b[0m\n",
      "\u001b[0;33m13:06:44 [base]\u001b[0m\u001b[0;35m Load file `data/disp.3L.txt`\u001b[0m\n",
      "Column 0: Agam_start - Agam_end (Agam_3L:2-41963435)\n",
      "  Agam_3L .. 12 (12) features .. +\n",
      "Column 1: Aalb_start - Aalb_end (Aalb_3R:2-34682003)\n",
      "  Aalb_3R .. 8 (8) features .. -\n",
      "Column 2: Afun_start - Afun_end (Afun_3:45633829-93833897)\n",
      "  Afun_3 .. 9 (9) features .. +\n",
      "/Users/am60/malaria/malariagen/binder/conda/envs/75_synteny_plot/lib/python2.7/site-packages/matplotlib/font_manager.py:1333: UserWarning: findfont: Font family [u'sans-serif'] not found. Falling back to DejaVu Sans\n",
      "  (prop.get_family(), self.defaultFamily[fontext]))\n",
      "\u001b[0;33m13:06:45 [base]\u001b[0m\u001b[0;35m Figure saved to `data/blocks.3L.pdf` (2400px x 2100px)\u001b[0m\n",
      "\u001b[0;33m13:06:46 [base]\u001b[0m\u001b[0;35m Load file `data/comb.2R.bed`\u001b[0m\n",
      "\u001b[0;33m13:06:46 [base]\u001b[0m\u001b[0;35m Load file `data/blocks.2R.txt`\u001b[0m\n",
      "\u001b[0;33m13:06:46 [base]\u001b[0m\u001b[0;35m Load file `data/disp.2R.txt`\u001b[0m\n",
      "Column 0: Agam_start - Agam_end (Agam_2R:2-61545105)\n",
      "  Agam_2R .. 19 (19) features .. +\n",
      "Column 1: Aalb_20 - Aalb_end (Aalb_2R:2-51801758)\n",
      "  Aalb_2R .. 18 (18) features .. +\n",
      "Column 2: Afun_start - Afun_end (Afun_2:2-56047416)\n",
      "  Afun_2 .. 16 (16) features .. +\n",
      "/Users/am60/malaria/malariagen/binder/conda/envs/75_synteny_plot/lib/python2.7/site-packages/matplotlib/font_manager.py:1333: UserWarning: findfont: Font family [u'sans-serif'] not found. Falling back to DejaVu Sans\n",
      "  (prop.get_family(), self.defaultFamily[fontext]))\n",
      "\u001b[0;33m13:06:47 [base]\u001b[0m\u001b[0;35m Figure saved to `data/blocks.2R.pdf` (2400px x 2100px)\u001b[0m\n",
      "\u001b[0;33m13:06:49 [base]\u001b[0m\u001b[0;35m Load file `data/comb.X.bed`\u001b[0m\n",
      "\u001b[0;33m13:06:49 [base]\u001b[0m\u001b[0;35m Load file `data/blocks.X.txt`\u001b[0m\n",
      "\u001b[0;33m13:06:49 [base]\u001b[0m\u001b[0;35m Load file `data/disp.X.txt`\u001b[0m\n",
      "Column 0: Agam_56 - Agam_end (Agam_X:2-24393108)\n",
      "  Agam_X .. 8 (8) features .. +\n",
      "Column 1: Afun_start - Afun_end (Afun_X:2-17661987)\n",
      "  Afun_X .. 6 (6) features .. -\n",
      "/Users/am60/malaria/malariagen/binder/conda/envs/75_synteny_plot/lib/python2.7/site-packages/matplotlib/font_manager.py:1333: UserWarning: findfont: Font family [u'sans-serif'] not found. Falling back to DejaVu Sans\n",
      "  (prop.get_family(), self.defaultFamily[fontext]))\n",
      "\u001b[0;33m13:06:49 [base]\u001b[0m\u001b[0;35m Figure saved to `data/blocks.X.pdf` (2400px x 2100px)\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "for chrom, chrom_len in AGAM_CHROM_SIZES.items():\n",
    "    # get amplicons on chromosome\n",
    "    chrom_ampls = comb_data.loc[comb_data.chrom == 'Agam_'+ chrom, 'amplicon']\n",
    "    # subset data to chromosome\n",
    "    chrom_data = comb_data[comb_data.amplicon.isin(chrom_ampls)].copy()\n",
    "    # check that only one chromosome per species was retained\n",
    "    uniq_chr = chrom_data.chrom.drop_duplicates()\n",
    "    for sp in SP_CAPTIONS.keys():\n",
    "        sp_chroms = uniq_chr.str.startswith(sp)\n",
    "        assert sp_chroms.sum() <= 1, \\\n",
    "            'Multiple chromosomes observed for {} in {}: {}'.format(sp, chrom, uniq_chr[sp_chroms].to_list()) \n",
    "    # enlarge regions\n",
    "    corr_data = longer_allsp(chrom_data, chrom_len)\n",
    "    # add chromosome regions\n",
    "    corr_data = add_ce(corr_data, chrom, CHROM_ENDS[chrom])\n",
    "    # BED files\n",
    "    to_tsv(COMB_BED.format(chrom),\n",
    "           corr_data[['chrom','corr_start','corr_end','id','color']])\n",
    "    # multiway blocks\n",
    "    write_blocks(corr_data, chrom)\n",
    "    # display layout\n",
    "    write_disp(corr_data, chrom)\n",
    "    # run visualisation\n",
    "    ! python -m jcvi.graphics.synteny --notex \\\n",
    "        {BLOCKS.format(chrom)} \\\n",
    "        {COMB_BED.format(chrom)} \\\n",
    "        {DISP_CONF.format(chrom)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "Exception",
     "evalue": "Analysis ended",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mException\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-23-e26a42e254f7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mraise\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Analysis ended'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mException\u001b[0m: Analysis ended"
     ]
    }
   ],
   "source": [
    "raise Exception('Analysis ended')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sandbox - test visualisation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import jcvi.graphics.synteny as sntn\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! python -m jcvi.graphics.synteny # --notex test/test.blocks.txt test/test.bed test/test.disp.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "fig, ax = plt.subplots(1,1, figsize=(10,10))\n",
    "\n",
    "sntn.Synteny(fig, ax, \n",
    "        datafile=BLOCKS.format(chrom), \n",
    "        bedfile=COMB_BED.format(chrom), \n",
    "        layoutfile=DISP_CONF.format(chrom))\n",
    "#             switch=switch, tree=tree, extra_features=opts.extra,\n",
    "#             scalebar=opts.scalebar)\n",
    "fig.show();\n",
    "# Synteny plotting code parts use both fig and ax to draw various elements\n",
    "# so making subplots work requires quite a lot of changes to the jcvi module code"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
